---
title: "Alternate splicing across EL and recovery"
output: pdf_document
---

```{r setup, include=FALSE, echo=FALSE}
library(reshape2)
library(tidyverse)
library(limma)
library(edgeR) 
library(scatterplot3d)

rrgd_dir <- "C:/Users/u4667515/Documents/Labwork/PostDoc_PEB/Experiments/2018/Alternative splicing exploration/RRGD_AS"
```

#### Aim
Explore extent of alternative splicing in _RRGD_ and in *sal1* and *xrn2xrn3*.

#### Method
Remine existing PE mRNA-sequencing datasets using the Arabidopsis Thaliana Reference Transcript Dataset 2 (AtRTD2). 

See: 

- Calixto, C.P.G., Guo, W., James, A.B., Tzioutziou, N.A., Entizne, J.C., Panter, P.E., Knight, H., Nimmo, H., Zhang, R., and Brown, J.W.S. (2018). Rapid and dynamic alternative splicing impacts the Arabidopsis cold response transcriptome. Plant Cell: tpc.00177.2018.
- [topSpliceDGE](https://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)
- [GitHub featureCounts script](https://github.com/dtrain16/NGS-scripts/blob/master/RNA/RNAseq_featureCounts_AS.sh)
- [Additional info](https://www.biostars.org/p/321379/)

## Alternative splicing under RRGD

```{r RRGD DGEList}
data_path <- paste0(rrgd_dir)
countFiles <- dir(path = data_path, pattern = "RTD2.counts")

## Define sample groups with descriptive labels from filenames
sampleGroups <- sapply(strsplit(countFiles, "-"), function(l) l[1])

## The DGElist object
lbls <- sapply(strsplit(countFiles, "_RTD2"), function(l) l[1])

input <- data_frame(countFiles) %>%
  mutate(file_contents = map(countFiles, ~read_delim(file.path(data_path, .),  delim = '\t', skip=1))) %>%
  unnest() %>%
  select(-countFiles, -Chr, -End, -Strand) %>%
  gather(sample, raw_counts, -Geneid, -Start, -Length) %>%
  na.omit() %>%
  mutate(sample = sapply(strsplit(sample, "/"), function(l) l[3])) %>%
  dcast(formula = Geneid + Start + Length ~ sample, value.var = 'raw_counts', fun.aggregate = mean)
  
y <- DGEList(counts = input[4:ncol(input)], genes=input[1:3], group = sampleGroups, samples = lbls)

### Abundance filter (CPM > 1 in at least 3 samples)
keep <- rowSums(cpm(y) > 1) > 3
dge <- y[keep, ]

## Re-calculate lib size based on retained transcripts
dge$samples$lib.size <- colSums(dge$counts)

# Use sample groups to make design matrix
design <- model.matrix(~0 + sampleGroups)
colnames(design) <- unique(sampleGroups)
design

## TMM Normalization and dispersion estimates
dge <- calcNormFactors(dge, method = "TMM") %>%
  estimateDisp(design, verobse=TRUE, robust=TRUE)
  
print(dge)

# quick cleanup for memory
rm(list = c('y','input','keep'))
```

DGEList setup, now can start estimating dispersion and fitting negative binomial GLMs...

```{r}
### Biological coefficient of variation
plotBCV(dge)

### MDS plots
groups <- unique(as.character(sampleGroups))
n.reps <- length(sampleGroups)/length(groups)

mds <- plotMDS(dge, ndim=3, dim.plot = c(1,2), col = rep(rainbow(length(groups)), each = n.reps))

s3d <- scatterplot3d(x = mds$cmdscale.out[,1:3], main = "3-dimensional MDS", xlab = "dim 1", ylab = "dim 2", zlab = "dim 3", color = rep(rainbow(length(groups)), each = n.reps), type='h', pch=19, lwd=1.5)
text(s3d$xyz.convert(mds$cmdscale.out[,1:3]), labels=sampleGroups, cex=.75, pos=4)

## GLM
my.contrasts <- makeContrasts(rrgd.II - rrgd.I,
levels=design
)
print(my.contrasts)

qlfit <- glmQLFit(dge, design, robust=TRUE, dispersion=dge$trended.dispersion)

# plot genewise quasi-likelihood dispersion
plotQLDisp(qlfit)
```

The above plots look a bit weird, likely due to assigning counts at the exon level (ie. multiple rows for different exons at the same gene). Onto alternative splicing (ie. testing for altered exon usage) after fitting quasi-likelihood negative binomial GLM from `glmQLfit`.

```{r}
### alternative splicing
sp <- diffSpliceDGE(qlfit, geneid = "Geneid", exonid = "Start", contrast = my.contrasts[,"rrgd.II - rrgd.I"])
tt <- topSpliceDGE(sp, test="gene", n=dim(sp$gene.genes)[1], FDR = 0.05)

head(tt)

# example AS loci
plotSpliceDGE(sp, geneid = paste(tt$Geneid[1]))

# an alternative, seemingly weaker test
# sv <- spliceVariants(dge, geneID = dge$genes$Geneid, group = dge$samples$group, estimate.genewise.disp = T)$table
# sv <- subset(sv, sv$PValue < 0.05) # would need to add adjustment
```

Looks like there are 65 alternatively spliced loci, see if any overlap with RRGD loci.

```{r}
rrgd_loci <- read.csv(file = "~/Manuscripts/2016_RRGD/TPC2016-00828-LSBR2_Supplemental_Data_Sets_5.csv", skip = 6) %>%
  select(c(1,7)) %>%
  mutate(ID = as.character(paste(Locus.Identifier)))

head(rrgd_loci)

common <- Reduce(intersect, list(rrgd_loci$ID, tt$Geneid))

# what genes overlap -- Aaron to model over RRGD?
print(common)

# how many genes
length(common)

# 11/65 AS = 16.9%

# which cateogry transcripts are these?
table(rrgd_loci$Category[rrgd_loci$Locus.Identifier %in% common])

# e.g. loci
plotSpliceDGE(sp, geneid = paste(common[2]))
```

#### Conclusion
Unclear, but maybe something there? There appears to be alternative splicing after 30 min EL, however, these do not appear to strongly correlate wit loci that undergo rapid decay during recovery (RRGD loci)
