---
title: "Alternate splicing across EL and recovery"
output: pdf_document
---

```{r setup, include=FALSE}
library(reshape2)
library(tidyverse)
library(limma)
library(edgeR) 
library(scatterplot3d)

rrgd_dir <- "C:/Users/u4667515/Documents/Labwork/PostDoc_PEB/Experiments/2018/Alternative splicing exploration/RRGD_AS"

## Output table
getAttributeField <- function (x, field, attrsep = ";") {
     s = strsplit(x, split = attrsep, fixed = TRUE)
     sapply(s, function(atts) {
         a = strsplit(atts, split = "=", fixed = TRUE)
         m = match(field, sapply(a, "[", 1))
         if (!is.na(m)) {
             rv = a[[m]][2]
         }
         else {
             rv = as.character(NA)
         }
         return(rv)
     })
}

gffRead <- function(gffFile, nrows = -1) {
     cat("Reading ", gffFile, ": ", sep="")
     gff = read.table(gffFile, sep="\t", as.is=TRUE, quote="",
     header=FALSE, comment.char="#", nrows = nrows,
     colClasses=c("character", "character", "character", "integer",
"integer",
     "character", "character", "character", "character"))
     colnames(gff) = c("seqname", "source", "feature", "start", "end",
             "score", "strand", "frame", "attributes")
        cat("found", nrow(gff), "rows with classes:",
        paste(sapply(gff, class), collapse=", "), "\n")
     stopifnot(!any(is.na(gff$start)), !any(is.na(gff$end)))
     return(gff)
}

# read in Araport11 gff3
anno <- gffRead("~/Labwork/Protocols/Araport11-igv/Araport11_GFF3_genes_transposons.201606.gff")
```

#### Aim
Explore extent of alternative splicing in _RRGD_ and in *sal1* and *xrn2xrn3*.

#### Method
Remine existing PE mRNA-sequencing datasets using the Arabidopsis Thaliana Reference Transcript Dataset 2 (AtRTD2). 

See: 

- Calixto, C.P.G., Guo, W., James, A.B., Tzioutziou, N.A., Entizne, J.C., Panter, P.E., Knight, H., Nimmo, H., Zhang, R., and Brown, J.W.S. (2018). Rapid and dynamic alternative splicing impacts the Arabidopsis cold response transcriptome. Plant Cell: tpc.00177.2018.
- [topSpliceDGE](https://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)
- [GitHub featureCounts script](https://github.com/dtrain16/NGS-scripts/blob/master/RNA/RNAseq_featureCounts_AS.sh)
- [Additional info](https://www.biostars.org/p/321379/)

## Alternative splicing under RRGD

```{r RRGD DGEList, message=FALSE}
data_path <- paste0(rrgd_dir)
countFiles <- dir(path = data_path, pattern = "RTD2.counts")

## Define sample groups with descriptive labels from filenames
sampleGroups <- sapply(strsplit(countFiles, "-"), function(l) l[1])

## The DGElist object
lbls <- sapply(strsplit(countFiles, "_RTD2"), function(l) l[1])

input <- data_frame(countFiles) %>%
  mutate(file_contents = map(countFiles, ~read_delim(file.path(data_path, .),  delim = '\t', skip=1))) %>%
  unnest() %>%
  select(-countFiles, -Chr, -End, -Strand) %>%
  gather(sample, raw_counts, -Geneid, -Start, -Length) %>%
  na.omit() %>%
  dcast(formula = Geneid + Start + Length ~ sample, value.var = 'raw_counts', fun.aggregate = mean)
  
y <- DGEList(counts = input[4:ncol(input)], genes=input[1:3], group = sampleGroups, samples = lbls)

### Abundance filter (CPM > 1 in at least 3 samples)
keep <- rowSums(cpm(y) > 1) > 3
dge <- y[keep, ]

## Re-calculate lib size based on retained transcripts
dge$samples$lib.size <- colSums(dge$counts)

# Use sample groups to make design matrix
design <- model.matrix(~0 + sampleGroups)
colnames(design) <- unique(sampleGroups)

## TMM Normalization and dispersion estimates
dge <- calcNormFactors(dge, method = "TMM") %>%
  estimateDisp(design, verobse=TRUE, robust=TRUE)
  
print(dge)

# quick cleanup for memory
rm(list = c('y','input','keep'))

### MDS plots
groups <- unique(as.character(sampleGroups))
n.reps <- length(sampleGroups)/length(groups)

mds <- plotMDS(dge, ndim=3, dim.plot = c(1,2), col = rep(rainbow(length(groups)), each = n.reps))

s3d <- scatterplot3d(x = mds$cmdscale.out[,1:3], main = "3-dimensional MDS", xlab = "dim 1", ylab = "dim 2", zlab = "dim 3", color = rep(rainbow(length(groups)), each = n.reps), type='h', pch=19, lwd=1.5)
text(s3d$xyz.convert(mds$cmdscale.out[,1:3]), labels=sampleGroups, cex=.75, pos=4)

```

Fit negative binomial GLMs for the following comparisons (and view dispersion of all transcripts in this dataset):

- EL30 vs 0 [rrgd.II - rrgd.I] 
- EL60 vs 0 [rrgd.III - rrgd.I] 
- EL60R7.5 vs 0  [rrgd.V - rrgd.I]

```{r}
## GLM
my.contrasts <- makeContrasts(
  rrgd.II - rrgd.I, 
  rrgd.III - rrgd.I, 
  rrgd.V - rrgd.I,
  levels=design
)

print(my.contrasts)

qlfit <- glmQLFit(dge, design, robust=TRUE, dispersion=dge$trended.dispersion)

# plot genewise quasi-likelihood dispersion
plotQLDisp(qlfit)
```

The above plots look a bit weird, likely due to assigning counts at the exon level (ie. multiple rows for different exons at the same gene). Onto alternative splicing (ie. testing for altered exon usage) after fitting quasi-likelihood negative binomial GLM from `glmQLfit`.

```{r, message=FALSE}
## get AS numbers across timepoints
for(i in colnames(my.contrasts)){
  sp <- diffSpliceDGE(qlfit, geneid = "Geneid", exonid = "Start", contrast = my.contrasts[,i])
  tt <- topSpliceDGE(sp, test="gene", n=dim(sp$gene.genes)[1], FDR = 0.05)
  print(i)
  print(dim(tt))
}

### alternative splicing of specific timepoint
sp <- diffSpliceDGE(qlfit, geneid = "Geneid", exonid = "Start", contrast = my.contrasts[,"rrgd.II - rrgd.I"])
tt_II <- topSpliceDGE(sp, test="gene", n=dim(sp$gene.genes)[1], FDR = 0.05)

sp <- diffSpliceDGE(qlfit, geneid = "Geneid", exonid = "Start", contrast = my.contrasts[,"rrgd.III - rrgd.I"])
tt_III <- topSpliceDGE(sp, test="gene", n=dim(sp$gene.genes)[1], FDR = 0.05)

sp <- diffSpliceDGE(qlfit, geneid = "Geneid", exonid = "Start", contrast = my.contrasts[,"rrgd.V - rrgd.I"])
tt_V <- topSpliceDGE(sp, test="gene", n=dim(sp$gene.genes)[1], FDR = 0.05)

test <- list(tt_II$Geneid, tt_III$Geneid, tt_V$Geneid)
length(Reduce(intersect, test))

```

Looks like there are 36 alternatively spliced loci in common across these timepoints, see if are also RRGD loci.

```{r}
rrgd_loci <- read.csv(file = "~/Manuscripts/2016_RRGD/TPC2016-00828-LSBR2_Supplemental_Data_Sets_5.csv", skip = 6) %>%
  select(c(1,7)) %>%
  mutate(ID = as.character(paste(Locus.Identifier)))

common <- Reduce(intersect, list(rrgd_loci$ID, Reduce(intersect, test)))

# what genes overlap -- Aaron to model over RRGD?
print(common)

# which cateogry transcripts are these?
table(rrgd_loci$Category[rrgd_loci$Locus.Identifier %in% common])

par(mfrow = c(2,2))

for(i in unique(common)){
plotSpliceDGE(sp, geneid = paste(i))
}

par(mfrow = c(1,1))
```

#### Conclusion
While there appears to be light-induced AS, it does not appear to be linked to transcript stability (at a transcriptome scale). There may still be transcript-dependent effects.
