---
title: "Alternate splicing in SAL1-PAP-XRN pathway"
output: pdf_document
---

```{r setup, include=FALSE, echo=FALSE}
library(reshape2)
library(tidyverse)
library(limma)
library(edgeR) 
library(scatterplot3d)

sal1_dir <- "C:/Users/u4667515/Documents/Labwork/PostDoc_PEB/Experiments/2018/Alternative splicing exploration/sal1_AS"
```

#### Aim
Explore extent of alternative splicing in *sal1* and *xrn2xrn3*.

#### Method
Remine existing PE mRNA-sequencing datasets using the Arabidopsis Thaliana Reference Transcript Dataset 2 (AtRTD2). 

See: 

- Calixto, C.P.G., Guo, W., James, A.B., Tzioutziou, N.A., Entizne, J.C., Panter, P.E., Knight, H., Nimmo, H., Zhang, R., and Brown, J.W.S. (2018). Rapid and dynamic alternative splicing impacts the Arabidopsis cold response transcriptome. Plant Cell: tpc.00177.2018.
- [topSpliceDGE](https://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)
- [GitHub featureCounts script](https://github.com/dtrain16/NGS-scripts/blob/master/RNA/RNAseq_featureCounts_AS.sh)
- [Additional info](https://www.biostars.org/p/321379/)

#### Results

```{r sal1 DGEList}
data_path <- paste0(sal1_dir)
countFiles <- dir(path = data_path, pattern = "RTD2.counts")

## Define sample groups with descriptive labels from filenames
sampleGroups <- sapply(strsplit(countFiles, "-"), function(l) l[1])

## The DGElist object
lbls <- sapply(strsplit(countFiles, "_RTD2"), function(l) l[1])

input <- data_frame(countFiles) %>%
  mutate(file_contents = map(countFiles, ~read_delim(file.path(data_path, .),  delim = '\t', skip=1))) %>%
  unnest() %>%
  select(-countFiles, -Chr, -End, -Strand) %>%
  gather(sample, raw_counts, -Geneid, -Start, -Length) %>%
  na.omit() %>%
  dcast(formula = Geneid + Start + Length ~ sample, value.var = 'raw_counts', fun.aggregate = mean)
  
y <- DGEList(counts = input[4:ncol(input)], genes=input[1:3], group = sampleGroups, samples = lbls)

### Abundance filter (CPM > 1 in at least 3 samples)
keep <- rowSums(cpm(y) > 1) > 3
dge <- y[keep, ]

## Re-calculate lib size based on retained transcripts
dge$samples$lib.size <- colSums(dge$counts)

# Use sample groups to make design matrix
design <- model.matrix(~0 + sampleGroups)
colnames(design) <- unique(sampleGroups)

## TMM Normalization and dispersion estimates
dge <- calcNormFactors(dge, method = "TMM") %>%
  estimateDisp(design, verobse=TRUE, robust=TRUE)
  
print(dge)

# quick cleanup for memory
rm(list = c('y','input','keep'))
```

DGEList setup, now can start estimating dispersion and fitting negative binomial GLMs...

```{r}
### Biological coefficient of variation
plotBCV(dge)

### MDS plots
groups <- unique(as.character(sampleGroups))

mds <- plotMDS(dge, ndim=3, dim.plot = c(1,2))

s3d <- scatterplot3d(x = mds$cmdscale.out[,1:3], main = "3-dimensional MDS", xlab = "dim 1", ylab = "dim 2", zlab = "dim 3", type='h', pch=19, lwd=1.5)
text(s3d$xyz.convert(mds$cmdscale.out[,1:3]), labels=sampleGroups, cex=.75, pos=4)

## GLM
my.contrasts <- makeContrasts(alx8 - WT, xrn23 - WT,
levels=design
)
print(my.contrasts)

qlfit <- glmQLFit(dge, design, robust=TRUE, dispersion=dge$trended.dispersion)

# plot genewise quasi-likelihood dispersion
plotQLDisp(qlfit)
```

The above plots look a bit weird, likely due to assigning counts at the exon level (ie. multiple rows for different exons at the same gene). Onto alternative splicing (ie. testing for altered exon usage) after fitting quasi-likelihood negative binomial GLM from `glmQLfit`.

##### alx8

```{r}
### alternative splicing
sp <- diffSpliceDGE(qlfit, geneid = "Geneid", exonid = "Start", contrast = my.contrasts[,"alx8 - WT"])
tt <- topSpliceDGE(sp, test="gene", n=dim(sp$gene.genes)[1], FDR = 0.05)

## how many
head(tt)
dim(tt)

par(mfcol = c(2,2))
# plot some top hits
for(i in 1:8){
  plotSpliceDGE(sp, geneid = paste(tt$Geneid[i]))
}
par(mfcol = c(1,1))

## Output table
getAttributeField <- function (x, field, attrsep = ";") {
     s = strsplit(x, split = attrsep, fixed = TRUE)
     sapply(s, function(atts) {
         a = strsplit(atts, split = "=", fixed = TRUE)
         m = match(field, sapply(a, "[", 1))
         if (!is.na(m)) {
             rv = a[[m]][2]
         }
         else {
             rv = as.character(NA)
         }
         return(rv)
     })
}

gffRead <- function(gffFile, nrows = -1) {
     cat("Reading ", gffFile, ": ", sep="")
     gff = read.table(gffFile, sep="\t", as.is=TRUE, quote="",
     header=FALSE, comment.char="#", nrows = nrows,
     colClasses=c("character", "character", "character", "integer",
"integer",
     "character", "character", "character", "character"))
     colnames(gff) = c("seqname", "source", "feature", "start", "end",
             "score", "strand", "frame", "attributes")
        cat("found", nrow(gff), "rows with classes:",
        paste(sapply(gff, class), collapse=", "), "\n")
     stopifnot(!any(is.na(gff$start)), !any(is.na(gff$end)))
     return(gff)
}

# read in Araport11 gff3
anno <- gffRead("~/Labwork/Protocols/Araport11-igv/Araport11_GFF3_genes_transposons.201606.gff")

output <- subset(anno,anno$feature=='gene') %>%
  mutate(Name=getAttributeField(attributes, 'Name')) %>%
  mutate(description=getAttributeField(attributes, 'Note')) %>%
	mutate(type=getAttributeField(attributes, 'locus_type')) %>%
	mutate(primary=getAttributeField(attributes, 'full_name')) %>%
	mutate(alias=getAttributeField(attributes, 'Alias')) %>%
  select('Name','description','type','primary','alias') %>%
  subset(Name %in% tt$Geneid) %>%
  mutate(Nexons = tt$NExons[match(Name, tt$Geneid)]) %>%
  mutate(FDR = tt$FDR[match(Name, tt$Geneid)])

write.csv(output, file="alx8_AS_gene-test_0.05_FDR.csv", row.names = F)

```

#### xrn2xrn3
```{r}
### alternative splicing
sp <- diffSpliceDGE(qlfit, geneid = "Geneid", exonid = "Start", contrast = my.contrasts[,"xrn23 - WT"])
tt <- topSpliceDGE(sp, test="gene", n=dim(sp$gene.genes)[1], FDR = 0.05)

# how many
head(tt)
dim(tt)

par(mfcol = c(2,2))
# plot some top hits
for(i in 1:8){
  plotSpliceDGE(sp, geneid = paste(tt$Geneid[i]))
}
par(mfcol = c(1,1))

## output table

output <- subset(anno,anno$feature=='gene') %>%
  mutate(Name=getAttributeField(attributes, 'Name')) %>%
  mutate(description=getAttributeField(attributes, 'Note')) %>%
	mutate(type=getAttributeField(attributes, 'locus_type')) %>%
	mutate(primary=getAttributeField(attributes, 'full_name')) %>%
	mutate(alias=getAttributeField(attributes, 'Alias')) %>%
  select('Name','description','type','primary','alias') %>%
  subset(Name %in% tt$Geneid) %>%
  mutate(Nexons = tt$NExons[match(Name, tt$Geneid)]) %>%
  mutate(FDR = tt$FDR[match(Name, tt$Geneid)])

write.csv(output, file="xrn2xrn3_AS_gene-test_0.05_FDR.csv", row.names = F)
```

#### Conclusion
Looks like a substantial number of AS events in *alx8*, but not so much in *xrn2xrn3*.


