---
title: "Experiment 555 RNA half-lives under differing conditions"
author: "Diep R Ganguly & Aaron B Smith"
date: "9 Aug 2021"
output: html_document
---

### Aim

Determine RNA half-lives measured, using cordycepin treatment paired with RNA-sequencing, in plants under control ("normal light", NL), high-light (HL), and recovery (Rec) conditions. 

```{r}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE)
options(stringsAsFactors = FALSE)

## packages
library(tidyverse)
library(tximport)
library(pheatmap)
library(RUVSeq)
#library(WGCNA)
#library(clusterSim)
library(factoextra)
library(patchwork)
library(RColorBrewer)

## install WGCNA and dependencies
# install.packages(c("matrixStats", "Hmisc", "splines", "foreach", "doParallel", "fastcluster", "dynamicTreeCut", "survival"))
# if (!requireNamespace("BiocManager", quietly = TRUE))
# 	install.packages("BiocManager")
# 	BiocManager::install(version = "3.13")
# BiocManager::install(c("GO.db", "preprocessCore", "impute")) 
# install.packages('WGCNA')

# gg theme
# source: https://ourcodingclub.github.io/tutorials/dataviz-beautification/
theme_niwot <- function(){
  theme_bw() +
    theme(
      text = element_text(family = "serif"), 
      axis.text = element_text(size = 12), 
      axis.title = element_text(size = 12),
      axis.line.x = element_line(color="black"),
      axis.line.y = element_line(color="black"),
          panel.grid.major.x = element_blank(),                                          
          panel.grid.minor.x = element_blank(),
          panel.grid.minor.y = element_blank(),
          panel.grid.major.y = element_blank(),  
          plot.margin = unit(c(1, 1, 1, 1), units = , "cm"),
          plot.title = element_text(size = 18, vjust = 1, hjust = 0),
          legend.text = element_text(size = 10),          
          legend.title = element_blank(),                              
          #legend.position = c(0.9, 0.9), 
          legend.key = element_blank(),
          legend.background = element_rect(color = "black", 
                                           fill = "transparent", 
                                           size = 2, linetype = "blank"))
}

## directories
quant <- "quant_files"
anno_dir <- "/home/diepg/ref_seqs/TAIR10"

## keyfile and unify timepoints across conditions
keyfile <- read_csv("exp_setup.csv") %>%
	mutate(Sample = as.factor(Sample)) %>%
	mutate(Treatment = factor(Treatment, levels = c("Mock", "Cordycepin"))) %>%
	mutate(Condition = factor(Condition, levels = c("NL", "EL", "Rec"))) %>%
	mutate(Timepoint = ifelse(Condition == "EL" & Timepoint == 60, 10,
		ifelse(Condition == "EL" & Timepoint == 70, 20,
		ifelse(Condition == "EL" & Timepoint == 80, 30,
		ifelse(Condition == "EL" & Timepoint == 90, 40,
		ifelse(Condition == "Rec" & Timepoint == 70, 20,
                ifelse(Condition == "Rec" & Timepoint == 80, 30,
                ifelse(Condition == "Rec" & Timepoint == 90, 40, Timepoint))))))))

## arrange sample information
metatable <- tibble(
  Sample = dir(path = quant),
) %>% arrange(as.numeric(paste0(metatable$Sample))) %>% left_join(keyfile, by="Sample")

new_lvls <- arrange(metatable, as.numeric(paste0(metatable$Sample)))$Sample
metatable$Sample <- factor(metatable$Sample, levels=new_lvls)

fls <- file.path(quant, metatable$Sample, "abundance.h5")
names(fls) <- metatable$Sample
all(file.exists(fls))

## transcript mapping info
tx2gene <- read_delim(file.path(anno_dir, "Arabidopsis_thaliana.TAIR10.51.gtf"), delim='\t', col_names = F, skip = 5) %>%
  filter(X3 == "transcript") %>%
  mutate(TXNAME = sapply(strsplit(X9, '"'), function(l) l[4])) %>%
  mutate(GENEID = sapply(strsplit(X9, '"'), function(l) l[2])) %>%
  dplyr::select(TXNAME,GENEID) %>%  unique

## generate TPM matrix
txi_gene <- tximport(files = fls,
                      tx2gene = tx2gene,
                      type = "kallisto",
                      countsFromAbundance = "lengthScaledTPM")

## abundance filter: TPM > 1 in all t0 samples
t0_ids <- subset(metatable, Timepoint == 10)
gene_tpm <- as.data.frame(txi_gene$abundance)
keep <- rowSums(gene_tpm[colnames(gene_tpm) %in% t0_ids$Sample]>1)>11
filtered <- round(txi_gene$counts[keep,], 0)

## Upper-quartile normalization
filtered_uq <- betweenLaneNormalization(filtered, which="upper")

clrs <- brewer.pal(4, "Set2")

pdf("Raw_counts_qc.pdf")
plotRLE(filtered, outline=FALSE, main = "Raw counts", ylim=c(-4, 4), col=clrs[metatable$Experiment])

pcomp1 <- prcomp(t(filtered), scale. = T)

fviz_pca_ind(pcomp1, repel = T, label="none",
        habillage=metatable$Experiment,
        addEllipses=TRUE, ellipse.type = "confidence")

fviz_pca_ind(pcomp1, repel = T, label="none",
        habillage=interaction(metatable$Treatment,metatable$Condition),
        addEllipses=TRUE, ellipse.type = "confidence")
dev.off()

pdf("UQ_normalized_counts_qc.pdf")

plotRLE(filtered_uq, outline=FALSE, main = "Upper-quartile correction", ylim=c(-4, 4), col=clrs[metatable$Experiment])

pcomp2 <- prcomp(t(filtered_uq), scale. = T)

fviz_pca_ind(pcomp2, repel = T, label="none", 
	habillage=metatable$Experiment, 
	addEllipses=TRUE, ellipse.type = "confidence")

fviz_pca_ind(pcomp2, repel = T, label="none", 
	habillage=interaction(metatable$Treatment,metatable$Condition), 
	addEllipses=TRUE, ellipse.type = "confidence")

dev.off()

## setup seq experiment object
set <- newSeqExpressionSet(
        counts=as.matrix(filtered),
        normalizedCounts=as.matrix(filtered_uq),
        phenoData = data.frame(metatable[2:5], row.names=metatable$Sample)
)

## Fit negative binomial GLM regression on UQ-normalized counts 
design <- model.matrix(~0 + metatable$Treatment * metatable$Condition * metatable$Timepoint,
	data=pData(set))
y <- DGEList(counts = counts(set), group = metatable$Sample)
y$samples <- left_join(y$samples, metatable, by=c("group"="Sample"))
y <- calcNormFactors(y, method = 'upperquartile')
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)

## fit NB-GLM on covariates of interest
fit <- glmFit(y, design)

## compute residuals
res <- residuals(fit, type="deviance")

# RUVr normalization - use all detected genes to estimate the factors of unwanted variation
genes <- rownames(filtered)
set_ruvr <- RUVr(set, genes, k=1, res)

#### sample clustering
my_grps <- dplyr::select(metatable, -Title, -Fastq, -checksum) %>% column_to_rownames("Sample") %>%
	mutate(Experiment = as.factor(paste(Experiment))) %>%
	mutate(Timepoint = as.factor(paste(Timepoint)))

pheatmap(as.matrix(cor(gene_tpm)),
        annotation_col = my_grps,
        filename = "2D_uncorrected_sample_clustering.pdf",
        scale = "none",
        angle_col = 90,
	show_rownames = F,
	fontsize_col = 4,
	fontsize = 6,
	cutree_cols = 3,
	cutree_rows = 3
)

## perform empirical-bayes adjustment
eblm <- empiricalBayesLM(
	t(gene_tpm),
	removedCovariates = metatable$Experiment,
	retainedCovariates = cbind(metatable$Timepoint,metatable$Condition,metatable$Timepoint),
	fitToSamples = )

pheatmap(as.matrix(cor(t(eblm$adjustedData))),
        annotation_col = my_grps,
        filename = "2D_corrected_sample_clustering.pdf",
        scale = "none",
        angle_col = 90,
        show_rownames = F,
        fontsize_col = 4,
        fontsize = 6,
        cutree_cols = 3,
        cutree_rows = 3
)


## re-scale adjusted values with unitization (zero minimum)
dat.eblm <- clusterSim::data.Normalization(x = t(eblm$adjustedData), 
	type = "n4", #unitization with zero minimum 
	normalization="row")

## perform PCA
pcomp1 <- prcomp(t(gene_tpm), scale. = T)
p1 <- fviz_eig(pcomp1)
p2 <- fviz_pca_ind(pcomp1, repel = T, label="none", habillage=metatable$Experiment, addEllipses=TRUE, ellipse.type = "confidence")

pcomp2 <- prcomp(dat.eblm, scale. = T)
p3 <- fviz_eig(pcomp2)
p4 <- fviz_pca_ind(pcomp2, repel = T, label="none", habillage=metatable$Experiment, addEllipses=TRUE, ellipse.type = "confidence")

pdf("EB-correction.pdf")
p1 + p2 + p3 + p4 + plot_annotation(title = 'Moderated EB correction (before and after)')
dev.off()

p5 <- fviz_pca_ind(pcomp2, repel = T, label="none", habillage=metatable$Timepoint, addEllipses=TRUE, ellipse.type = "confidence")
p6 <- fviz_pca_ind(pcomp2, repel = T, label="none", habillage=metatable$Treatment, addEllipses=TRUE, ellipse.type = "confidence")
p7 <- fviz_pca_ind(pcomp2, repel = T, label="none", habillage=metatable$Condition, addEllipses=TRUE, ellipse.type = "confidence")
p8 <- fviz_pca_ind(pcomp2, repel = T, label="none", habillage=metatable$Experiment, addEllipses=TRUE, ellipse.type = "confidence")

pdf("multiPCA_corrected_v1.pdf")
p5 + p6 + p7 + p8
dev.off()

p9 <- fviz_pca_ind(pcomp2, repel = T, label="none", 
	habillage=interaction(metatable$Treatment,metatable$Condition,metatable$Timepoint), 
	addEllipses=TRUE, ellipse.type = "confidence")

pdf("multiPCA_corrected_v2.pdf")
p8 + p9
dev.off()

# transform data for modelling
input <-as.data.frame(t(dat.eblm)) %>%
	rownames_to_column('ID') %>%
	gather(key=Sample, value=tpm, -ID) %>%
	left_join(keyfile, by='Sample') %>%
	dplyr::select(-Title, -Fastq, -checksum)

## seperate three time-courses by condition (i.e. NL, HL, or Rec)
## Note, HL and Rec share the same 10 minute reference sample
## scale expression values relative to t10
rel_scale <- subset(input, Timepoint == 10) %>%
        group_by(ID, Timepoint, Treatment, Condition) %>%
	mutate(tpm = tpm) %>%
	summarise(avg_tpm = mean(tpm))

## scale values relative to t0
d1 <- subset(input, Condition == "NL") %>%
	mutate(rel_tpm = rel_scale$avg_tpm[match(interaction(ID, Treatment, Condition), interaction(rel_scale$ID, rel_scale$Treatment, rel_scale$Condition))]) %>%
	mutate(rel_abundance = tpm/rel_tpm)

d2 <- subset(input, Condition == "EL" | Condition == "Rec") %>%
	mutate(rel_tpm = rel_scale$avg_tpm[match(interaction(ID, Treatment, Condition), interaction(rel_scale$ID, rel_scale$Treatment, rel_scale$Condition))]) %>%
        mutate(rel_abundance = tpm/rel_tpm)

###########################
######### Exploratory smoothed fitting
###########################

### look at trends for random genes
sample_ids <- sample(unique(input$ID), size = 10, replace = FALSE)

pdf("explore_smoothjitter_NL.pdf")
subset(d1, ID %in% sample_ids ) %>%
	group_by(ID, Treatment, Timepoint) %>% 
	summarise(avg_tpm = mean(tpm)) %>%
	ggplot(aes(x = Timepoint, y = avg_tpm, colour = Treatment)) + 
		geom_line() +
		geom_point(data=subset(d1, ID %in% sample_ids ), 
			aes(x=Timepoint, y=tpm, colour=Treatment), size=1, alpha=0.4) + 
		facet_wrap(~ID) + 
		coord_cartesian(ylim= c(0,1)) +
		scale_y_continuous(name = "Relative abundance", breaks=c(0,1)) + 
		scale_x_continuous(name = "Time (min)", breaks=c(10,20,30)) + 
		theme_niwot() +
		theme(legend.position = "bottom")
dev.off()

pdf("explore_smoothjitter_ELvsRec.pdf")
subset(d2, ID %in% sample_ids ) %>%
        group_by(ID, Treatment, Condition, Timepoint) %>%
        summarise(avg_tpm = mean(tpm)) %>%
        ggplot(aes(x = Timepoint, y = avg_tpm, colour=Treatment, linetype=Condition)) +
                geom_line() +
		geom_point(data=subset(d2, ID %in% sample_ids ),
                     aes(x=Timepoint, y=tpm, colour=Treatment, shape=Condition), size=1, alpha=0.4) +
                facet_wrap(~ID) +
		coord_cartesian(ylim= c(0,1)) +
                scale_y_continuous(name = "Relative abundance", breaks=c(0,1)) +
                scale_x_discrete(name = "Time (min)", breaks=c(10,20,30,40)) +
                theme_niwot() +
		theme(legend.position = "bottom")
dev.off()

### trial NL models

pdf("gompertz_NL_trial.pdf")

par(mfrow = c(3, 3))

# plot and fit to 9 random genes
for (i in sample(unique(d1$ID), size = 9, replace = FALSE)) {
	a <- subset(d1, ID %in% i)
	
	a1 <- subset(a, Treatment == "Mock")
	a2 <- subset(a, Treatment == "Cordycepin")

	# ERROR HANDLING for initial fit
possibleError <- tryCatch(tmp <- getInitial(Y ~ SSlogis(X, Asym, xmid, scal), 
        data = a), error = function(e) e)	

possibleError <- tryCatch(tmp <- getInitial(Y ~ SSgompertz(X, Asym, b2, b3), data = a1), 
	error = function(e) e)
    
	if (!inherits(possibleError, "error")) {
		tmp <- getInitial(Y ~ SSgompertz(X, Asym, b2, b3), data = a1)
    }
    
    # ERROR HANDLING for final fit
    possibleError <- tryCatch(fit <- gnls(Y ~ SSgompertz(X, Asym, b2, b3), data = a1, 
        start = tmp), error = function(e) e)
    
    if (!inherits(possibleError, "error")) {
        # REAL WORK
        fit <- gnls(Y ~ SSgompertz(X, Asym, b2, b3), data = a1, start = tmp)
        yfitted <- predict(fit, list(X = xNew))

	plot(Y ~ X, a, ylab = "Gompertz Mock", xlab = "Time (min)")
        lines(xNew, yfitted, col = "salmon")
        mtext(paste(i), font = 3, padj = -1)

    }
}

dev.off()



