---
title: "Experiment 555 RNA half-lives under differing conditions"
author: "Aaron, Diep"
date: "21 May 2021"
output: html_document
---


### Aim

Determine RNA half-lives measured, using cordycepin treatment paired with RNA-sequencing, in plants under no stress, light-stress, and recovery. 

```{r}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE)
options(stringsAsFactors = FALSE)

library(tidyverse)
library(tximport)
library(lme4)
library(lmerTest)

## file directories
quant <- "Kallisto_quant"
anno_dir <- "~/ref_seqs/TAIR10/"

## keyfile and unify timepoints across conditions
keyfile <- read_csv("exp_setup.csv") %>%
	mutate(Sample = as.factor(Sample)) %>%
	mutate(Condition = factor(Condition, levels = c("NL", "EL", "Rec"))) %>%
	mutate(Timepoint = ifelse(Condition == "EL" & Timepoint == 60, 10,
		ifelse(Condition == "EL" & Timepoint == 70, 20,
		ifelse(Condition == "EL" & Timepoint == 80, 30,
		ifelse(Condition == "EL" & Timepoint == 90, 40,
		ifelse(Condition == "Rec" & Timepoint == 70, 20,
                ifelse(Condition == "Rec" & Timepoint == 80, 30,
                ifelse(Condition == "Rec" & Timepoint == 90, 40, Timepoint))))))))


key <- NULL
for(i in unique(keyfile$Condition)){
	a <- subset(keyfile, Condition == "i")
	

## arrange sample information
metatable <- tibble(
  title = dir(path = quant),
  Sample = as.factor(sapply(strsplit(title, "_"), function(l) l[2]))
) %>% left_join(keyfile, by="Sample")

fls <- file.path(quant, metatable$title, "2_quant", "abundance.h5")
names(fls) <- metatable$Sample
all(file.exists(fls))

## transcript mapping info
tx2gene <- read_delim(file.path(anno_dir, "Arabidopsis_thaliana.TAIR10.50.gtf"), delim='\t', col_names = F, skip = 5) %>%
  filter(X3 == "transcript") %>%
  mutate(TXNAME = sapply(strsplit(X9, '"'), function(l) l[4])) %>%
  mutate(GENEID = sapply(strsplit(X9, '"'), function(l) l[2])) %>%
  select(TXNAME,GENEID) %>%  unique

## generate TPM matrix
txi_gene <- tximport(files = fls,
                      tx2gene = tx2gene,
                      type = "kallisto",
                      countsFromAbundance = "no")

## TPM matrix of genes detected at TPM > 1 in at least 3 samples
gene_tpm <- txi_gene$abundance[rowSums(txi_gene$abundance > 1) > 3, ]

## transform data for modelling
input <- as.data.frame(gene_tpm) %>%
	rownames_to_column('ID') %>%
	gather(key=Sample, value=tpm, -ID) %>%
	left_join(keyfile, by='Sample') %>%
	mutate(log2_tpm = log2(tpm))



## build model
mdl <- lmer(formula = tpm ~ Treatment * Timepoint * Condition + (1 | Experiment), data = input)

for(i in unique(gene_tpm$ID)){

}

## generate gene-level TPM matrix
fls_NL <- fls[names(fls) %in% subset(metatable, Condition == "NL")$title]
txi_NL <- tximport(files = fls_NL, tx2gene = tx2gene, type = "kallisto", countsFromAbundance = "no")

##---> TPM matrix for downstream analyses
tpm_NL <- txi_NL$abundance[rowSums(txi_NL$abundance > 1) > 3, ] ## TPM > 1 in at least 3
colnames(tpm_NL) <- sapply(strsplit(colnames(tpm_NL), "_"), function(l) l[2])

tpm_NL <- as.data.frame(tpm_NL) %>% 
  rownames_to_column('ID') %>%
  gather(key=Sample, value=tpm, -ID) %>%
  left_join(keyfile, by='Sample')

mdl <- lmer(formula = log2(tpm) ~ Treatment + Timepoint + (1 | Experiment), data = subset(tpm_NL, ID == "AT1G01010"))

ggplot() +
  geom_point(aes(x=Timepoint, y=log2(tpm), colour=Treatment), data = subset(tpm_NL, ID == "AT1G01010"))


```



