---
title: "Experiment 555 RNA half-lives under differing conditions"
author: "Diep R Ganguly & Aaron B Smith"
date: "20 Aug 2021"
output: html_document
---

### Aim

Determine RNA half-lives measured, using cordycepin treatment paired with RNA-sequencing, in plants under control ("normal light", NL), high-light (HL), and recovery (Rec) conditions. 

```{r}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE)
options(stringsAsFactors = FALSE)

## packages
library(tidyverse)
library(tximport)
library(pheatmap)
library(RUVSeq)
library(factoextra)
library(patchwork)
library(RColorBrewer)
library(lme4)
library(piecewiseSEM)

# gg theme
# source: https://ourcodingclub.github.io/tutorials/dataviz-beautification/
theme_niwot <- function(){
  theme_bw() +
    theme(
      text = element_text(family = "serif"), 
      axis.text = element_text(size = 12), 
      axis.title = element_text(size = 12),
      axis.line.x = element_line(color="black"),
      axis.line.y = element_line(color="black"),
          panel.grid.major.x = element_blank(),                                          
          panel.grid.minor.x = element_blank(),
          panel.grid.minor.y = element_blank(),
          panel.grid.major.y = element_blank(),  
          plot.margin = unit(c(1, 1, 1, 1), units = , "cm"),
          plot.title = element_text(size = 18, vjust = 1, hjust = 0),
          legend.text = element_text(size = 10),          
          legend.title = element_blank(),                              
          #legend.position = c(0.9, 0.9), 
          legend.key = element_blank(),
          legend.background = element_rect(color = "black", 
                                           fill = "transparent", 
                                           size = 2, linetype = "blank"))
}

## directories
quant <- "quant_files"
anno_dir <- "/home/diepg/ref_seqs/TAIR10"

## metadata and normalize timepoints from time of infiltration
metatable <- read_csv("exp_setup.csv") %>%
	mutate(Sample = as.factor(Sample)) %>%
	mutate(Treatment = factor(Treatment, levels = c("Mock", "Cordycepin"))) %>%
	mutate(Condition = factor(Condition, levels = c("NL", "EL", "Rec"))) %>%
	mutate(Timepoint = ifelse(Condition == "EL" & Timepoint == 60, 10,
		ifelse(Condition == "EL" & Timepoint == 70, 20,
		ifelse(Condition == "EL" & Timepoint == 80, 30,
		ifelse(Condition == "EL" & Timepoint == 90, 40,
		ifelse(Condition == "Rec" & Timepoint == 70, 20,
                ifelse(Condition == "Rec" & Timepoint == 80, 30,
                ifelse(Condition == "Rec" & Timepoint == 90, 40, Timepoint))))))))

new_lvls <- arrange(metatable, as.numeric(paste0(metatable$Sample)))$Sample
metatable$Sample <- factor(metatable$Sample, levels=new_lvls)

fls <- file.path(quant, metatable$Sample, "abundance.h5")
names(fls) <- metatable$Sample
all(file.exists(fls))

## transcript mapping info
tx2gene <- read_delim(file.path(anno_dir, "Arabidopsis_thaliana.TAIR10.51.gtf"), delim='\t', col_names = F, skip = 5) %>%
  filter(X3 == "transcript") %>%
  mutate(TXNAME = sapply(strsplit(X9, '"'), function(l) l[4])) %>%
  mutate(GENEID = sapply(strsplit(X9, '"'), function(l) l[2])) %>%
  dplyr::select(TXNAME,GENEID) %>%  unique

##########################
###### Data import and abundance filter
##########################

## generate TPM matrix
txi_gene <- tximport(files = fls,
                      tx2gene = tx2gene,
                      type = "kallisto",
                      countsFromAbundance = "lengthScaledTPM")

## abundance filter: TPM > 1 in all t10 values
t0_ids <- subset(metatable, Timepoint == 10)$Sample
gene_tpm <- as.data.frame(txi_gene$abundance)
keep <- rowSums(gene_tpm[colnames(gene_tpm) %in% t0_ids]>1)>11

## counts with integer structure
filtered <- round(txi_gene$counts[keep,], 0)

#################
#### RUV - remove unwanted variance (batch) (Risso et al 2014)
#################

## setup seq experiment object
set <- newSeqExpressionSet(counts=as.matrix(filtered),
	phenoData = data.frame(metatable[2:5], row.names=metatable$Sample)
)

## Fit negative binomial GLM regression 
design <- model.matrix(~Treatment * Condition * Timepoint, data=pData(set))

y <- DGEList(counts = counts(set), group = metatable$Sample)
y$samples <- left_join(y$samples, metatable, by=c("group"="Sample"))
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)

## fit NB-GLM on covariates of interest
fit <- glmFit(y, design)

## compute residuals
res <- residuals(fit, type="deviance")

## RUVr normalization
genes <- rownames(filtered)
set1 <- RUVr(set, genes, k=1, res)

## RUVs normalization
x <- interaction(metatable$Treatment,metatable$Timepoint,metatable$Condition)
differences <- makeGroups(x)
set2 <- RUVs(set, genes, k=1, differences)

## colours
clrs <- brewer.pal(6, "Set2")

pdf("RLEplots.pdf")
plotRLE(as.matrix(gene_tpm[keep,]), outline=FALSE, main = "TPM", 
	ylim=c(-4, 4), col=clrs[metatable$Treatment])
plotRLE(cpm(normCounts(set1)), outline=FALSE, main = "RUVr(set, genes, k=1, res)", 
	ylim=c(-4, 4), col=clrs[metatable$Treatment])
plotRLE(cpm(normCounts(set2)), outline=FALSE, main = "RUVs(set, genes, k=1, differences)", 
	ylim=c(-4, 4), col=clrs[metatable$Treatment])
dev.off()

pdf("PCAplots.pdf")
pcomp1 <- prcomp(t(as.matrix(gene_tpm[keep,])), scale. = T)
pcomp2 <- prcomp(t(cpm(normCounts(set1))), scale. = T)

fviz_pca_ind(pcomp1, repel = T, label="none", title="TPM",
        habillage=metatable$Experiment, 
	xlim=c(-200,200), ylim=c(-200,200),
        addEllipses=TRUE, ellipse.type = "confidence")

fviz_pca_ind(pcomp2, repel = T, label="none", title = "RUVr",
        habillage=pData(set1)$Experiment, addEllipses=TRUE,
	xlim=c(-200,200), ylim=c(-200,200),
        ellipse.type = "confidence")

fviz_pca_ind(pcomp1, repel = T, label="none", title="TPM",
        habillage=interaction(metatable$Treatment,metatable$Condition),
	xlim=c(-200,200), ylim=c(-200,200),
        addEllipses=TRUE, ellipse.type = "confidence")

fviz_pca_ind(pcomp2, repel = T, label="none", title = "RUVr",
        habillage=interaction(pData(set1)$Treatment, pData(set1)$Condition),
	xlim=c(-200,200), ylim=c(-200,200),
        addEllipses=TRUE, ellipse.type = "confidence")
dev.off()

## Extract corrected CPM values
input_p1 <- as.data.frame(cpm(normCounts(set1))) %>%
	rownames_to_column('ID') %>%
	gather(Sample, cpm, -ID) %>%
	left_join(metatable, by="Sample") %>%
	dplyr::select(-Title, -Fastq, -checksum, -Sample)

input_p2 <- subset(input_p1, Timepoint == 10 & Condition == "EL") %>%
	mutate(Condition = gsub(pattern="EL", replacement="Rec", x=Condition))

input <- rbind(input_p1, input_p2)

#####################
### Normalization - T0 and decay factor
####################

## Convert CPM to fractional decreases from T10 for each condition
## Note, HL and Rec share the same 10 minute reference sample
t10 <- subset(input, Timepoint == 10) %>%
        group_by(ID, Timepoint, Treatment, Condition) %>%
	mutate(cpm = as.numeric(paste0(cpm))) %>%
	summarise(mean_cpm = mean(cpm))

## scale values relative to t10 and apply decay factor normalization to cordycepin-treated samples
d1 <- subset(input, Treatment == "Cordycepin") %>%
        mutate(t10_cpm = t10$mean_cpm[match(interaction(ID, Treatment, Condition),
                interaction(t10$ID, t10$Treatment, t10$Condition))]) %>%
        mutate(frac_cpm = cpm/t10_cpm) %>%
        select(-cpm, -t10_cpm, -Treatment)

## make list of stable genes based on expression (cpm) and variance (cov)
x <- as.data.frame(cpm(normCounts(set1))) %>%
        mutate(std = apply(X = ., MARGIN = 1, FUN = sd)) %>%
        mutate(avg = rowMeans(., na.rm = T)) %>%
        mutate(cov = std/avg)

cv <- kmeans(x$cov, centers=10, iter.max= 100, algorithm="Lloyd")
cv2 <- kmeans(log2(x$avg+0.01), centers=10, iter.max= 100, algorithm="Lloyd")

## 20th percentile COV
genes1 <- names(subset(cv$cluster, as.data.frame(cv$cluster)[,1] == order(cv$centers)[2]))
## 90th percentile abundance
genes2 <- names(subset(cv2$cluster, as.data.frame(cv2$cluster)[,1] == order(cv2$centers)[10]))
## reference genes
refgenes <- Reduce(intersect, list(genes1,genes2))

pdf("refgenes.pdf")
ggplot(data=subset(x, !(rownames(x) %in% refgenes))) + aes(y=cov, x=log2(avg+0.01)) +
        geom_jitter(size=0.5, alpha=0.3) +
	geom_jitter(data=subset(x, rownames(x) %in% refgenes), colour='salmon') +
        theme_niwot() +
        scale_y_continuous(name = "Coefficient of variation") +
        scale_x_continuous(name = "log2 [TPM+0.01]")
dev.off()

## calclate decay factor for cordycepin-treated samples at each timepoint and condition
normFactors_mean <- subset(d1, ID %in% refgenes) %>% 
	group_by(Timepoint, Condition) %>%
	summarise(decay_fac = mean(frac_cpm))

## apply decay normalization and filter any genes with a fold-increase after T10
d2 <- mutate(d1, decay_fac = normFactors_mean$decay_fac[match(interaction(Timepoint,Condition), 
	interaction(normFactors_mean$Timepoint, normFactors_mean$Condition))]) %>%
	mutate(decaynorm_frac_cpm = frac_cpm/decay_fac) %>%
	mutate(test = ifelse(decaynorm_frac_cpm > 1.5 & Timepoint != 10, yes = 1, no = 0))

gene_filter <- subset(d2, test == 1) %>% select(ID) %>% unique

normData <- select(d2, -frac_cpm, -test, -decay_fac) %>%
	subset(!(ID %in% gene_filter$ID)) %>%
	mutate(log_e = log(decaynorm_frac_cpm + 0.01))

###########################
######### Data exploration 
###########################

### look at trends for random genes
sample_ids <- sample(unique(normData$ID), size = 16, replace = FALSE)

pdf("data_explore.pdf")
subset(normData, ID %in% sample_ids ) %>%
	group_by(ID, Timepoint, Condition) %>%
	summarise(relative_abundance = mean(decaynorm_frac_cpm), 
		reps = sum(!(is.na(decaynorm_frac_cpm))),
		SE = sd(decaynorm_frac_cpm)/sqrt(reps)) %>%
	ggplot(aes(x = Timepoint, y = relative_abundance, col = Condition)) + 
		geom_line(linetype='dashed', size=0.5, alpha=0.6)+
		geom_point(size=0.8) + 
		geom_errorbar(aes(ymin=relative_abundance-SE, ymax=relative_abundance+SE), width=.8) +
		facet_wrap(~ID) + 
		coord_cartesian(ylim= c(0,1.25)) +
		scale_y_continuous(name = "Relative abundance", breaks=c(0,1)) + 
		scale_x_continuous(name = "Time (min)", breaks=c(10,20,30,40)) + 
		scale_colour_manual(values = c('NL'="gray60","EL"="salmon","Rec"="royalblue")) +
		theme_niwot() +
		theme(legend.position = "bottom")
dev.off()

## setup df for model-based predictions
n.preds=42
r <- range(normData$Timepoint)
exp_r <- unique(normData$Experiment)
newdf <- data.frame(Timepoint=seq(r[1],r[2],length.out = n.preds)) %>%
	mutate(Experiment = rep(exp_r, n.preds/length(unique(normData$Experiment) )))


out <- NULL
for( i in unique(normData$ID)){

tmp <- subset(normData, ID == i)

	for( c in unique(tmp$Condition)){
	
	tmp1 <- subset(tmp, Condition == c)
	possibleError <- tryCatch(fit <- lmer(log_e ~ Timepoint + (1| Experiment), data=tmp1), 
		error = function(e) e)

	tmp_out <- NULL
	if ( !inherits(possibleError, "error")) {
		fit <- lmer(log_e ~ Timepoint + (1|Experiment), data=tmp1)

		rsq <- rsquared(fit)$Conditional
		k <- -coefficients(fit)$Experiment$Timepoint[1]	
		bic <- BIC(fit)
		half_life <- log(2)/k
		a1 <- data.frame(geneID=i, bic, rsq, k, half_life, Condition=c)	
		tmp_out <- rbind(tmp_out, a1)
	}	
}

out <- rbind(out, tmp_out)

}


## produce sample plots

	pred1 <- predict(mdl1, newdf, re.form=NA)

	ggplot()+
		theme_niwot() +
		geom_point(aes(x=Timepoint, y=log_e), data=tmp1)+
		geom_line(aes(x=newdf$Timepoint, y=pred1), colour="gray60", linetype='dashed')

}





