---
title: "RNA decay protein alignments"
author: "Diep"
header-includes:
   - \usepackage{texshade}
date: "13 May 2019"
output:
  pdf_document: 
    latex_engine: xelatex
---

## Aim

Explore amino acid sequences of RNA binding proteins to inform protein engineering methodology (e.g. saturation mutagenesis of conserved residues or fusion proteins of functional domains) for *Bio-capacitors*.

## Methods

### Phylogenetic analyses
All analyses were performed in *R* (v3.5.0, R Core Team 2018). Amino acid sequences of candidate orthologs (isoform 1) were collated from NCBI and Uniprot. Unannotated orthologs were identified using PBLAST at CNGBdb. These sequences were aligned with the *Clustal Omega* algorithm, which pairs guide tree construction (*mBed*) with pairwise HMM alignment (*HHalign*) facilitating scalability for large datasets ( [Sievers et al 2011](https://www.ncbi.nlm.nih.gov/pubmed/21988835?dopt=Abstract) ) using the *msa* package (Bodenhofer et al 2015). Construction of bootstrapped maximum likelihood trees were performed using the *phagnorn* package ( [Schliep et al 2011](https://academic.oup.com/bioinformatics/article/27/4/592/198887) ). The starting tree was first generated by computing pairwise distances (*dist.ml*, model = "Blosum62") followed by UPGMA clustering (*upgma*). The likelihood for this tree was computed (*pml*) and optmized using the likelihood ratchet for tree optimization (*optim.pml*, model = "Blosum62", rearrangement = "ratchet") (Vos 2003).


```{r, include=FALSE}
knitr::opts_chunk$set(tidy=TRUE, echo=FALSE, message=FALSE, warning=FALSE)

# libraries
library(tidyverse)
library(msa)
library(phangorn)
suppressPackageStartupMessages(library(dendextend))

fldir <- file.path(getwd(), "FASTAs")
fqs <- tibble(fls = dir(fldir))
```

## Results

### 5'-3' exoribonuclease 1 (XRN1)

Look at conserved regions described to be important for XRN activity (Chang *et al.* 2011; Braun *et al.* 2012);
Jinek *et al.* 2011).

```{r}
# read in fasta file of protein sequences
my_seqs <- readAAStringSet(filepath = file.path(fldir, fqs$fls[5]), format = 'fasta')

# Clustal Omega algorith for aa alignment
my_align <- msaClustalOmega(my_seqs)
```

```{r, results='asis', eval=FALSE}
## show alignment
msaPrettyPrint(x=my_align, 
               y=c(1,1000), 
               output = "asis", 
               showLogo="none",
               consensusColor="BlueRed",
               shadingMode="similar",
               shadingColors = "blues")
```

```{r}
## convert to phyDat object
my_align2 <- as.phyDat(my_align, type="AA")

## compute pairwise distances using BLOSUM62 scoring matrix
d <- as.matrix(dist.ml(my_align2, model = "Blosum62"))

## generate starting tree with upgma
treeUPGMA <- upgma(d)

## compute maximum likelihood
fit <- pml(treeUPGMA, data = my_align2)

## branch-swapping to optimize tree parameters based on Blosum62 model with the likelihood ratchet (Vos RA, 2003)
opfit <- optim.pml(fit, model = "Blosum62", rearrangement = "ratchet")

anova(fit, opfit)

par(mar=c(1,1,1,1))
plot.phylo(opfit$tree, main = "XRN1 phylogeny", cex=.8, label.offset = 0.02)

# # bs = bootstrap.pml(fit62, bs=100, optNni=TRUE, multicore=TRUE)
# plotBS(fit$tree, BS, type = "phylogram")

```

### Pumilio (PUM) proteins

The PUF protein family are RBPs  found across Eukaryotes. They are involved in post-transcriptional control (RNA stability and translation) whereby they bind AREs through sequence recognition by the *Pumilio homology domain* (PUM-HD) consisting of eight tandem copes of a 36 amino acid motif (PUM repeat) (Wang *et al.* 2001; Cheong C-G, Hall TMT. 2006). We can utilize the PUM-HD for *Bio-capacitor* targeting. 

The PUM-HD of human Pumilio homolog 1 has been used to direct the PIN domain of HsSmg6 for targeted endonuclealytic activity (Choudhury *et al.* 2012). Therefore, *HsPUM1* can be used as a reference point. The Wt version should target *Drosophila* NRE1 5'-UGUAUAUA-3' (Cheong *et al.* 2006; Francischini *et al.* 2009). APUM1-6 are group I PUM proteins and should act similarly to *Drosophila* and Human PUM1 (Francischini *et al.* 2009). The resolved PUM-HD, from HsPUM1, used for engineering with residues 828â€“1,176 (Choudhury *et al.* 2012). Drosophila PUM-HD occurs from residues 1,091 - 1,428.

#### Clustal omega multiple sequence alignment

```{r}
# read in fasta file of protein sequences
my_seqs <- readAAStringSet(filepath = file.path(fldir, fqs$fls[3]), format = 'fasta')

## perform clustal omega alignment
my_align <- msaClustalOmega(my_seqs, verbose = FALSE)
```

Show alignment of PUM-HD domain:

```{r, results='asis'}
## show alignment
msaPrettyPrint(x=my_align, showNames = "right", showNumbering = "left", 
               y=c(1450,1850), 
               output = "asis", 
               shadingMode="similar",
               furtherCode=c("\\showruler{1}{top}"))
```

#### Phylogenetic tree

```{r}
## convert to phyDat object
my_align2 <- as.phyDat(my_align, type="AA")

## compute pairwise distances using BLOSUM62 scoring matrix
d <- as.matrix(dist.ml(my_align2, model = "Blosum62"))

## generate starting tree with upgma
treeUPGMA <- upgma(d)

## compute maximum likelihood
fit <- pml(treeUPGMA, data = my_align2)

## branch-swapping to optimize tree parameters based on Blosum62 model with the likelihood ratchet (Vos RA, 2003)
opfit <- optim.pml(fit, model = "Blosum62", rearrangement = "ratchet",
)

anova(fit, opfit)

par(mar=c(1,1,1,1))
plot.phylo(opfit$tree, main = "Pumilio proteins", cex=.8, label.offset = 0.02)

# # bs = bootstrap.pml(fit62, bs=100, optNni=TRUE, multicore=TRUE)
# plotBS(fit$tree, BS, type = "phylogram")

# circlize_dendrogram(fit)

```

