---
title: "mRNA fate dynamics during light stress and recovery"
author: "Diep R Ganguly & Aaron B Smith"
date: "20 Aug 2021"
output: html_document
---

### Aim

Determine RNA half-lives measured, using cordycepin treatment paired with RNA-sequencing, in plants under control ("normal light", NL), high-light (HL), and recovery (Rec) conditions. 

```{r}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE)
options(stringsAsFactors = FALSE)

## packages
library(tidyverse)
library(ggExtra)
library(tximport)
library(pheatmap)
library(RUVSeq)
library(factoextra)
library(patchwork)
library(RColorBrewer)
library(lme4)
library(lmerTest)
library(piecewiseSEM)
library(dabestr)
library(janitor)
library(VennDiagram)

# gg theme
# source: https://ourcodingclub.github.io/tutorials/dataviz-beautification/
theme_niwot <- function(){
  theme_bw() +
    theme(
      text = element_text(family = "serif"), 
      axis.text = element_text(size = 12), 
      axis.title = element_text(size = 12),
      axis.line.x = element_line(color="black"),
      axis.line.y = element_line(color="black"),
          panel.grid.major.x = element_blank(),                                          
          panel.grid.minor.x = element_blank(),
          panel.grid.minor.y = element_blank(),
          panel.grid.major.y = element_blank(),  
          plot.margin = unit(c(0.5, .5, .5, .5), units = , "cm"),
          plot.title = element_text(size = 18, vjust = 1, hjust = 0),
          legend.text = element_text(size = 10),          
          legend.title = element_blank(),                              
          #legend.position = c(0.9, 0.9), 
          legend.key = element_blank(),
          legend.background = element_rect(color = "black", 
                                           fill = "transparent", 
                                           size = 2, linetype = "blank"))
}

## Output table
getAttributeField <- function (x, field, attrsep = ";") {
     s = strsplit(x, split = attrsep, fixed = TRUE)
     sapply(s, function(atts) {
         a = strsplit(atts, split = "=", fixed = TRUE)
         m = match(field, sapply(a, "[", 1))
         if (!is.na(m)) {
             rv = a[[m]][2]
         }
         else {
             rv = as.character(NA)
         }
         return(rv)
     })
}

gffRead <- function(gffFile, nrows = -1) {
     gff = read.table(gffFile, sep="\t", as.is=TRUE, quote="",
     header=FALSE, comment.char="#", nrows = nrows,
     colClasses=c("character", "character", "character", "integer", "integer",
     "character", "character", "character", "character"))
     colnames(gff) = c("seqname", "source", "feature", "start", "end",
             "score", "strand", "frame", "attributes")
     stopifnot(!any(is.na(gff$start)), !any(is.na(gff$end)))
     return(gff)
}

## directories
quant <- "quant_files"
anno_dir <- "~/ref_seqs/"
anno <- gffRead(file.path(anno_dir, "Arabidopsis_thaliana.TAIR10.51.gff3")) %>%
  subset(feature == "gene") %>% 
  mutate(ID = getAttributeField(attributes, "gene_id")) %>%
  mutate(name = getAttributeField(attributes, "Name")) %>%
  mutate(description = getAttributeField(attributes, "description")) %>%
  select(ID, name, description)

## metadata and normalize timepoints from time of infiltration
metatable <- read_csv("exp_setup.csv") %>%
	mutate(Sample = as.factor(Sample)) %>%
	mutate(Treatment = factor(Treatment, levels = c("Mock", "Cordycepin"))) %>%
	mutate(Condition = factor(Condition, levels = c("NL", "EL", "Rec"))) %>%
	mutate(Timepoint = ifelse(Condition == "EL" & Timepoint == 60, 10,
		ifelse(Condition == "EL" & Timepoint == 70, 20,
		ifelse(Condition == "EL" & Timepoint == 80, 30,
		ifelse(Condition == "EL" & Timepoint == 90, 40,
		ifelse(Condition == "Rec" & Timepoint == 70, 20,
                ifelse(Condition == "Rec" & Timepoint == 80, 30,
                ifelse(Condition == "Rec" & Timepoint == 90, 40, Timepoint))))))))

## RRGD genes
rrgd <- read_csv("TPC2016-00828-LSBR2_Supplemental_Data_Sets_5.csv", skip = 6) %>%
  select(-c(3:6,33:ncol(.))) %>%
  clean_names %>% subset(rrgd_loci == "RRGD")

new_lvls <- arrange(metatable, as.numeric(paste0(metatable$Sample)))$Sample
metatable$Sample <- factor(metatable$Sample, levels=new_lvls)

fls <- file.path(quant, metatable$Sample, "abundance.h5")
names(fls) <- metatable$Sample
all(file.exists(fls))

## transcript mapping info
tx2gene <- read_delim(file.path(anno_dir, "Arabidopsis_thaliana.TAIR10.51.gtf"), delim='\t', col_names = F, skip = 5) %>%
  filter(X3 == "transcript") %>%
  mutate(TXNAME = sapply(strsplit(X9, '"'), function(l) l[4])) %>%
  mutate(GENEID = sapply(strsplit(X9, '"'), function(l) l[2])) %>%
  dplyr::select(TXNAME,GENEID) %>%  unique

##########################
###### Data import and abundance filter
##########################

## generate TPM matrix
txi_gene <- tximport(files = fls,
                      tx2gene = tx2gene,
                      type = "kallisto",
                      countsFromAbundance = "lengthScaledTPM")

## abundance filter: TPM > 0.5 in >50% of t10 samples
t0_ids <- subset(metatable, Timepoint == 10)$Sample
gene_tpm <- as.data.frame(txi_gene$abundance)
keep <- rowSums(gene_tpm[colnames(gene_tpm) %in% t0_ids]>0.5)>6

## counts with integer structure
filtered <- round(txi_gene$counts[keep,], 0)

#################
#### RUV - remove unwanted variance (batch) (Risso et al 2014)
#################

## setup seq experiment object
set <- newSeqExpressionSet(counts=as.matrix(filtered),
	phenoData = data.frame(metatable[2:5], row.names=metatable$Sample)
)

## Fit negative binomial GLM regression 
design <- model.matrix(~Treatment * Condition * Timepoint, data=pData(set))

y <- DGEList(counts = counts(set), group = metatable$Sample)
y$samples <- left_join(y$samples, metatable, by=c("group"="Sample"))
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)

## fit NB-GLM on covariates of interest
fit <- glmFit(y, design)

## compute residuals
res <- residuals(fit, type="deviance")

## RUVr normalization
genes <- rownames(filtered)
set1 <- RUVr(set, genes, k=1, res)

## RUVs normalization
#x <- interaction(metatable$Treatment,metatable$Timepoint,metatable$Condition)
#differences <- makeGroups(x)
#set2 <- RUVs(set, genes, k=1, differences)

## colours
clrs <- brewer.pal(6, "Set2")

pdf("RUV_corrected_abundance.pdf")
plotRLE(as.matrix(gene_tpm[keep,]), outline=FALSE, main = "TPM", 
	ylim=c(-4, 4), col=clrs[metatable$Treatment])
plotRLE(cpm(normCounts(set1)), outline=FALSE, main = "RUVr", 
	ylim=c(-4, 4), col=clrs[metatable$Treatment])
#plotRLE(cpm(normCounts(set2)), outline=FALSE, main = "RUVs", 
#	ylim=c(-4, 4), col=clrs[metatable$Treatment])
dev.off()

## Extract corrected CPM values
input_p1 <- as.data.frame(cpm(normCounts(set1))) %>%
	rownames_to_column('ID') %>%
	gather(Sample, cpm, -ID) %>%
	left_join(metatable, by="Sample") %>%
	dplyr::select(-Title, -Fastq, -checksum, -Sample)

input_p2 <- subset(input_p1, Timepoint == 10 & Condition == "EL") %>%
	mutate(Condition = gsub(pattern="EL", replacement="Rec", x=Condition))

input <- rbind(input_p1, input_p2)

#####################
### BCA plot for reference genes
####################

## make list of stable genes based on expression (cpm) and variance (cov)
x <- subset(input, Treatment == "Mock") %>%
	group_by(ID) %>%
	summarise(mean_cpm = mean(cpm), cov = sd(cpm)/mean_cpm) %>%
	arrange(mean_cpm)

## <25th percentile COV & >98th percentile mean CPM
refgenes <- subset(x, cov < quantile(x$cov, c(0.25)) & mean_cpm > quantile(x$mean_cpm, c(0.98)))

y1 <- subset(input, Treatment == "Cordycepin") %>%
        group_by(ID) %>%
        summarise(mean_cpm = mean(cpm), cov = sd(cpm)/mean_cpm) %>%
        arrange(mean_cpm)

pdf("BCV-plot-refgenes.pdf")
ggplot(data=x) + aes(y=cov, x=log2(mean_cpm+0.01)) +
        geom_jitter(data=subset(x, !(ID %in% refgenes$ID)), aes(color="genes"), size=0.5, alpha=0.3) +
	geom_jitter(data=subset(x, ID %in% refgenes$ID), aes(color='reference genes')) +
	ggtitle('Mock-treated samples') +
	labs(color="Legend") +
	scale_color_manual(values=c("genes"="black","reference genes"="salmon")) +
	theme_niwot() +
	coord_cartesian(ylim=c(0,5), xlim=c(-5,15)) +
        scale_y_continuous(name = "Coefficient of variation") +
        scale_x_continuous(name = "log2 [CPM+0.01]")

ggplot(data=y1) + aes(y=cov, x=log2(mean_cpm+0.01)) +
        geom_jitter(data=subset(y1, !(ID %in% refgenes$ID)), aes(color="genes"), size=0.5, alpha=0.3) +
        geom_jitter(data=subset(y1, ID %in% refgenes$ID), aes(color='reference genes')) +
        ggtitle('Cordycepin-treated samples') +
	labs(color="Legend") +
        scale_color_manual(values=c("genes"="black","reference genes"="salmon")) +
	theme_niwot() +
	coord_cartesian(ylim=c(0,5), xlim=c(-5,15)) +
        scale_y_continuous(name = "Coefficient of variation") +
        scale_x_continuous(name = "log2 [CPM+0.01]")

dev.off()

#####################
### t10 and decay factor normalization
####################

## compute mean cpm at t10 n.b. EL and Rec share t10
t10 <- subset(input, Timepoint == 10) %>%
        group_by(ID, Timepoint, Treatment, Condition) %>%
        mutate(cpm = as.numeric(paste0(cpm))) %>%
        summarise(ref_cpm = mean(cpm))

## Convert CPM to fractional decreases relative to t10
d1 <- subset(input, Treatment == "Cordycepin") %>%
        mutate(t10_cpm = t10$ref_cpm[match(interaction(ID, Treatment, Condition),
                interaction(t10$ID, t10$Treatment, t10$Condition))]) %>%
        mutate(frac_cpm = cpm/t10_cpm) %>%
        select(-cpm, -t10_cpm, -Treatment)

## calculate mean fold-change in reference genes in cordycepin-treated samples per condition
normFactors_mean <- subset(d1, ID %in% refgenes$ID) %>% 
	group_by(Timepoint, Condition, Experiment) %>%
	summarise(decay_fac = mean(frac_cpm))

## apply decay normalization and filter any genes with a fold-increase after T10 (all conditions)
d2 <- mutate(d1, decay_fac = normFactors_mean$decay_fac[match(interaction(Timepoint,Condition,Experiment), interaction(normFactors_mean$Timepoint, normFactors_mean$Condition, normFactors_mean$Experiment))]) %>%
	mutate(decaynorm_frac_cpm = frac_cpm/decay_fac) %>%
	mutate(test = ifelse(decaynorm_frac_cpm >= 1.5 & Timepoint != 10, yes = 1, no = 0)) %>%
	mutate(test = ifelse(is.na(decaynorm_frac_cpm) == T, 2, test))

gene_filter <- subset(d2, test == 1 | is.na(test) == T | test == 2) %>% select(ID) %>% unique

## fractional CPM
normData2 <- select(d2, -decaynorm_frac_cpm, -test, -decay_fac) %>%
        subset(!(ID %in% gene_filter$ID)) %>%
        mutate(log_e = log(frac_cpm + 0.01))

## fractional CPM normalized by decay factor
normData <- select(d2, -frac_cpm, -test, -decay_fac) %>%
	subset(!(ID %in% gene_filter$ID)) %>%
	mutate(log_frac_cpm = log(decaynorm_frac_cpm + 0.01))

## decay normalization and filter genes with a fold-increase after T10 (NL only)
d3 <- subset(d1, Condition == "NL") %>%
	mutate(decay_fac = normFactors_mean$decay_fac[match(interaction(Timepoint,Condition,Experiment), interaction(normFactors_mean$Timepoint, normFactors_mean$Condition, normFactors_mean$Experiment))]) %>%
        mutate(decaynorm_frac_cpm = frac_cpm/decay_fac) %>%
        mutate(test = ifelse(decaynorm_frac_cpm >= 1.5 & Timepoint != 10, yes = 1, no = 0)) %>%
	mutate(test = ifelse(is.na(decaynorm_frac_cpm) == T, 2, test))

gene_filter <- subset(d3, test == 1 | is.na(test) == T | test == 2) %>% select(ID) %>% unique

## fractional CPM normalized by decay factor for NL samples
normData_NL <- select(d3, -frac_cpm, -test, -decay_fac) %>%
        subset(!(ID %in% gene_filter$ID)) %>%
        mutate(log_frac_cpm = log(decaynorm_frac_cpm + 0.01))

## loci to undergo half-life modelling
length(unique(normData_NL$ID))
length(unique(normData$ID))

###########################
#### evaluate decrease in RNA pool and normalize
###########################

x2 <- subset(metatable, Treatment=="Cordycepin") %>%
	mutate(test=ifelse(test=duplicated(.[2:5]), 1, 0))

x1 <- subset(normData, !(Condition == "Rec" & Timepoint == 10)) %>% 
	arrange(Condition, Timepoint) %>%
	mutate(test=ifelse(test=duplicated(.[1:4]), 1, 0)) %>%
	mutate(Sample=x2$Sample[match(interaction(Timepoint,Condition,Experiment,test), interaction(x2$Timepoint,x2$Condition,x2$Experiment,x2$test))]) %>%
	na.omit() %>%
	select(Sample,ID,decaynorm_frac_cpm) %>%
	mutate(Sample = as.character(Sample)) %>%
	mutate(Sample = factor(Sample, levels=unique(Sample))) %>%
	spread(Sample, decaynorm_frac_cpm) %>%
	column_to_rownames('ID')

x3 <- tibble(sample = colnames(x1)) %>%
	left_join(x2, by=c('sample'='Sample'))

x4 <- subset(normData2, !(Condition == "Rec" & Timepoint == 10)) %>%
        arrange(Condition, Timepoint) %>%
        mutate(test=ifelse(test=duplicated(.[1:4]), 1, 0)) %>%
        mutate(Sample=x2$Sample[match(interaction(Timepoint,Condition,Experiment,test), interaction(x2$Timepoint,x2$Condition,x2$Experiment,x2$test))]) %>%
        na.omit() %>%
        select(Sample, ID, frac_cpm) %>%
        mutate(Sample = as.character(Sample)) %>%
        mutate(Sample = factor(Sample, levels=unique(Sample))) %>%
        spread(Sample, frac_cpm) %>%
        column_to_rownames('ID')

tmp <- gene_tpm[colnames(gene_tpm) %in% colnames(x1)] %>%
        rownames_to_column('ID') %>%
        gather(sample, tpm, -ID) %>%
        left_join(x2, by=c('sample'='Sample')) %>%
        mutate(test=ifelse(test=duplicated(.[1:4]), 1, 0)) %>%
        na.omit()  %>%
        select(sample, ID, tpm) %>%
        mutate(sample = as.character(sample)) %>%
        mutate(sample = factor(sample, levels=unique(x3$sample))) %>%
        spread(sample, tpm) %>%
        column_to_rownames('ID')

pdf("RNA_pool_test.pdf")
clrs <- c(brewer.pal(9, "Set1"), brewer.pal(8, "Set2"))

plotRLE(as.matrix(tmp), 
	outline=FALSE, ylim=c(-1, 1), main = "TPM (all genes)", ylab = "relative log-expression",
        col=clrs[as.factor(interaction(x3$Condition,x3$Timepoint))])

plotRLE(as.matrix(subset(tmp, rownames(tmp) %in% rownames(x4))),
        outline=FALSE, ylim=c(-1, 1), main = "TPM", ylab = "relative log-expression",
        col=clrs[as.factor(interaction(x3$Condition,x3$Timepoint))])

plotRLE(as.matrix(x4), outline=FALSE, ylim=c(-1, 1), main = "Fractional CPM", ylab = "relative log-expression",
        col=clrs[as.factor(interaction(x3$Condition,x3$Timepoint))])

plotRLE(as.matrix(x1), outline=FALSE, ylim=c(-1, 1), main = "Fractional CPM + decay normalization", ylab = "relative log-expression", col=clrs[as.factor(interaction(x3$Condition,x3$Timepoint))])

dev.off()

###########################
#### evaluate cordycepin treatment based on DGE testing with RUVr adjusted counts
###########################

### Setup edgeR object for DGE testing
y2 <- as.data.frame(normCounts(set1))
y2 <- DGEList(counts = y2, group=metatable$Sample)
y2$samples <- left_join(y2$samples,metatable, by=c("group"="Sample"))
y2$samples$sampleGroup <- paste(y2$samples$Treatment,y2$samples$Timepoint,y2$samples$Condition, sep='_')
design <- model.matrix(~0 + sampleGroup, data=y2$samples)
colnames(design) <- sapply(strsplit(colnames(design), 'sampleGroup'), function(l) l[2])

y2 <- estimateDisp(y2, design, robust=TRUE)
fit <- glmQLFit(y2, design, robust=TRUE)

my.contrasts <- makeContrasts(
mock_NL = Mock_30_NL - Mock_10_NL,
mock_EL = Mock_40_EL - Mock_10_EL,
mock_Rec = Mock_40_Rec - Mock_10_EL,
cordycepin_NL = Cordycepin_30_NL - Cordycepin_10_NL,
cordycepin_EL = Cordycepin_40_EL - Cordycepin_10_EL,
cordycepin_Rec = Cordycepin_40_Rec - Cordycepin_10_EL,
levels=design
)

pdf("scatterplot_reps.pdf")
ggplot(data.frame(y2$counts)) + aes(x=X48, y=X85) + 
	geom_point() +
	theme_niwot() +
	geom_abline(color='blue', linetype = 'dashed')+
	coord_cartesian(xlim=c(0,10000), ylim=c(0,10000))
dev.off()


dge <- NULL

for(i in colnames(my.contrasts)){
  res <- glmQLFTest(fit, contrast = my.contrasts[,i])
  tt <- topTags(res, adjust.method = "fdr", sort.by = "none", p.value=1, n=dim(res)[1])$table
  tt$contrast <- paste(i)
  tt$ID <- rownames(tt)
  tt <- select(tt, ID, logCPM, contrast, logFC, FDR)
  dge <- rbind(dge,tt)
}

dge <- mutate(dge, DGE=ifelse(logFC > 0 & FDR < 0.05, yes = "Up-regulated", no=
                      ifelse(logFC < 0 & FDR < 0.05, yes='Down-regulated', no="Non-DGE")))

dge_out <- dge %>% subset(DGE != "Non-DGE") %>% left_join(anno, by=c("ID"))
write_csv(dge_out, "cordycepin_dge.csv")

pdf("PCAplots_cordycepin_RNA.pdf")

mds <- plotMDS(y2, ndim=3, plot=F, gene.selection = "pairwise")

mds_df <- data.frame(Sample = colnames(mds$distance.matrix.squared), 
	x1 = as.numeric(mds$x), 
	y1 = as.numeric(mds$y)) %>%
	left_join(metatable, by="Sample") %>%
	mutate(Timepoint = as.factor(paste(Timepoint)))

ggplot(data = mds_df) + 
	aes(x=x1, y=y1, colour=Condition, shape=Treatment) +
	geom_point() +
	facet_wrap(~Timepoint) + 
	scale_x_continuous(name = "Dim1") +
	scale_y_continuous(name = "Dim2") +
	theme_niwot()

dev.off()


pdf("DGE_testing.pdf")

ggplot(data=subset(dge, DGE != "Non-DGE")) + aes(x=logCPM, y=logFC, group=contrast, colour=DGE) +
  geom_jitter(size=0.8) +
  theme_niwot() +
  geom_hline(yintercept = 0, linetype='dashed', col='salmon', alpha=0.5) +
  facet_wrap(~contrast) +
  scale_y_continuous(name = "log2 fold-change", limits = c(-8,8)) +
  scale_x_continuous(name = "log2 CPM") +
  scale_colour_manual(values = c('Down-regulated'="royalblue","Up-regulated"="khaki1","Non-DGE"="lightgrey"))

dev.off()

###########################
#### Modelling half-life - unstressed; NL
###########################

out1 <- NULL
for( i in unique(normData_NL$ID)){
	
	tmp <- subset(normData_NL, ID == i)
	tmp_out <- NULL
	possibleError <- tryCatch(fit <- lmer(log_frac_cpm ~ Timepoint + (1| Experiment), data=tmp), error = function(e) e)

	if ( !inherits(possibleError, "error")) {
		fit <- lmer(log_frac_cpm ~ Timepoint + (1|Experiment), data=tmp)
		rsq <- rsquared(fit)$Conditional
		pval <- anova(fit)$'Pr(>F)'
                aic <- AIC(fit)
                k <- -coefficients(fit)$Experiment$Timepoint[1]
                halflife <- log(2)/k
                a1 <- data.frame(geneID=i, rsq, pval, aic, k, halflife, Condition=tmp$Condition[1])
                tmp_out <- rbind(tmp_out, a1)
        }

out1 <- rbind(out1, tmp_out)
}

## filter genes
out_NL <- subset(out1, pval < 0.05 & k > 0)

p1 <- ggplot(out_NL) + aes(x=k) +
        geom_histogram(position="identity", bins=20) +
        theme_niwot() +
        theme(legend.position = 'none') +
        scale_x_continuous(name = "k-decay")

p2 <- ggplot(out_NL) + aes(x=rsq, y=halflife) +
        geom_point() +
        theme_niwot() +
	#coord_cartesian(xlim=c(0,1)) +
        scale_x_continuous(name = "R-sq")+
        scale_y_continuous(name = "Half-life (min)")

p3 <- ggplot(out_NL) + aes(x=pval, y=halflife) +
        geom_point() +
        theme_niwot() +
        scale_x_continuous(name = "p-value")+
        scale_y_continuous(name = "Half-life (min)")

pdf("model_parameters_NL.pdf")
p1 / ggMarginal(p2, type = "histogram") + ggMarginal(p3, type = "histogram")
dev.off()

###########################
#### Modelling half-life - all conditions
###########################

out2 <- NULL

### subset to check if removing T40 effects HL/REC modelling
# normData <- subset(normData, Timepoint != 40)

for( i in unique(normData$ID)){

tmp <- subset(normData, ID == i)

	tmp_out <- NULL
	
	for( c in  unique(tmp$Condition)){
	
	tmp1 <- subset(tmp, Condition == c)
	possibleError <- tryCatch(fit <- lmer(log_frac_cpm ~ Timepoint + (1| Experiment), data=tmp1), 
		error = function(e) e)

	if ( !inherits(possibleError, "error")) {
		fit <- lmer(log_frac_cpm ~ Timepoint + (1|Experiment), data=tmp1)

		rsq <- rsquared(fit)$Conditional
                pval <- anova(fit)$'Pr(>F)'
		aic <- AIC(fit)
		k <- -coefficients(fit)$Experiment$Timepoint[1]	
		halflife <- log(2)/k
		a1 <- data.frame(geneID=i, rsq, pval, aic, k, halflife, Condition=c)	
		tmp_out <- rbind(tmp_out, a1)
	}	
}

out2 <- rbind(out2, tmp_out)
out2$Condition <- factor(out2$Condition, levels=c("NL","EL","Rec"))
}

## filter genes
out2_filt <- unique(subset(out2, pval > 0.05 | k < 0)$geneID)
out_all <- subset(out2, !(geneID %in% out2_filt)) %>%
	mutate(log_halflife = log2(halflife))

p1 <- ggplot(out_all) + aes(x=k, colour=Condition, fill=Condition) +
        geom_histogram(position="identity", bins=20, alpha=0.5) +
        theme_niwot() +
	theme(legend.position = 'none') +
        scale_x_continuous(name = "k-decay")

p2 <- ggplot(out_all) + aes(x=rsq, y=halflife, color=Condition) +
        geom_point() +
        theme_niwot() +
	#facet_wrap(~Condition) +
	#coord_cartesian(xlim=c(0,1)) +
	scale_x_continuous(name = "R-sq")+
        scale_y_continuous(name = "Half-life (min)")

p3 <- ggplot(out_all) + aes(x=pval, y=halflife, color=Condition) +
        geom_point() +
        theme_niwot() +
	#facet_wrap(~Condition) +
        scale_x_continuous(name = "p-value")+
        scale_y_continuous(name = "Half-life (min)")

pdf("model_parameters_all.pdf")
p1 / ggMarginal(p2, groupColour = TRUE, groupFill = TRUE, type = "histogram") + ggMarginal(p3, groupColour = TRUE, groupFill = TRUE, type = "histogram")
dev.off()

################
## Venn modelling results
################
a <- mutate(out2, test = ifelse(pval < 0.05 & k > 0, yes = 1, no =0)) %>%
	select(geneID, Condition, test) %>%
	spread(Condition, test)

pdf("modelling_conditions.pdf")
vennDiagram(a[2:4], mar=c(2,2,2,2), lwd=1, cex=1)
dev.off()


###########################
## Heatmap of fractional decrease
###########################

my_grps <- dplyr::select(metatable, -Title, -Fastq, -checksum, -Experiment, -Treatment) %>%
        column_to_rownames("Sample") %>%
        mutate(Timepoint = as.factor(paste(Timepoint)))

heatmap_input <- group_by(normData, ID, Timepoint, Condition) %>%
        summarise(rel_cpm = mean(decaynorm_frac_cpm)) %>% mutate(Treatment = "Cordycepin") %>%

	subset(ID %in% out_all$geneID) %>% ## subset to 3960 modelled genes

        mutate(Sample = metatable$Sample[match(interaction(Timepoint,Condition,Treatment), interaction(metatable$Timepoint,metatable$Condition,metatable$Treatment))]) %>%
        na.omit() %>% ungroup() %>% select(-Timepoint, -Condition, -Treatment) %>%
        spread(Sample, rel_cpm) %>% column_to_rownames('ID')

heatmap_input <- as.matrix(heatmap_input)

rowDistance = dist(heatmap_input, method = "euclidean")
rowCluster = hclust(rowDistance, method = "complete")

hmap <- pheatmap(heatmap_input,
        annotation_col = my_grps,
        cluster_rows=rowCluster,
	cluster_cols=F,
        filename = "heatmap_normalized_fractional_all.pdf",
        scale = "none",
        angle_col = 90,
	cutree_rows = 4,
        show_rownames = F,
        show_colnames = F,
        fontsize = 6
)

hmap_rows <- cutree(rowCluster, 4)
hmap_rows <- as.data.frame(hmap_rows[hmap$tree_row$order])
colnames(hmap_rows) <- "decay_cat"

out_all_decay <- mutate(out_all, decay_category = hmap_rows$decay_cat[match(geneID, rownames(hmap_rows))]) %>% 
	arrange(match(geneID, rownames(hmap_rows))) %>% 
	select(-log_halflife, -aic) %>% 
	mutate(RRGD = ifelse(geneID %in% rrgd$locus_identifier == T, yes = 1, no=0))

write_csv(out_all_decay, "allconditions_halflife.csv")

subset(hmap_rows, decay_cat == 3) %>% rownames_to_column("geneID") %>% write_csv("fast_decay_genes.csv")

subset(hmap_rows, decay_cat == 2) %>% rownames_to_column("geneID") %>% write_csv("slow_decay_genes.csv")

subset(hmap_rows, decay_cat == 4) %>% rownames_to_column("geneID") %>% write_csv("rapid_decay_genes.csv")

subset(hmap_rows, decay_cat == 1) %>% rownames_to_column("geneID") %>% write_csv("stable_decay_genes.csv")

###############################
## decay profile of sample genes
##############################

# test2 <- select(rrgd, locus_identifier) %>%
#	mutate(r_sq_nl = out_NL$rsq[match(locus_identifier, out_NL$geneID)]) %>%
#	na.omit() %>%
#	arrange(r_sq_nl)

# sample_ids <- sample(unique(subset(out_all, rsq>0.8)$geneID), size = 3, replace = FALSE)
# sample_ids <- c("AT1G74310","AT2G47180","AT3G25230")

model_ids <- hmap_rows %>% rownames_to_column('geneID') %>% left_join(out_all) %>%
	group_by(geneID, decay_cat) %>%
	summarise(mean_rsq = mean(rsq), std = sd(rsq)) %>%
	mutate(crit = mean_rsq/std)

sample_t1 <- subset(model_ids, decay_cat == 1) %>% arrange(desc(crit))
sample_t1 <- sample_t1[1:2,]

sample_t2 <- subset(model_ids, decay_cat == 2) %>% arrange(desc(crit))
sample_t2 <- sample_t2[1:2,]

sample_t3 <- subset(model_ids, decay_cat == 3) %>% arrange(desc(crit))
sample_t3 <- sample_t3[1:2,]

sample_t4 <- subset(model_ids, decay_cat == 4) %>% arrange(desc(crit))
sample_t4 <- sample_t4[1:2,]

sample_ids <- rbind(sample_t1, sample_t2, sample_t3, sample_t4) %>% select(geneID)

pdf("sample_model_fit.pdf", paper="a4")
test <- subset(normData, ID %in% sample_ids$geneID ) %>%
        group_by(ID, Timepoint, Condition) %>%
        summarise(relative_abundance = mean(decaynorm_frac_cpm),
                reps = sum(!(is.na(decaynorm_frac_cpm))),
                SE = sd(decaynorm_frac_cpm)/sqrt(reps)) %>%
	mutate(rsq = out2$rsq[match(interaction(ID,Condition), interaction(out2$geneID,out2$Condition))]) %>% 
	mutate(halflife=out2$halflife[match(interaction(ID,Condition), interaction(out2$geneID,out2$Condition))])

ggplot(data=test, aes(x = Timepoint, y = relative_abundance, col = Condition)) +
	geom_line(linetype='dashed', size=0.5, alpha=0.4)+
        geom_point(size=0.6) +
        geom_errorbar(aes(ymin=relative_abundance-SE, ymax=relative_abundance+SE),width=.8)+
        facet_grid(Condition~ID) +
        coord_cartesian(ylim= c(0,1.2)) +
        scale_y_continuous(name = "Normalized relative abundance", breaks=c(0,0.5,1)) +
        scale_x_continuous(name = "Time post-infiltration (min)", breaks=c(10,20,30,40)) +
        theme_niwot() +
	geom_hline(yintercept = 0.5, linetype='dashed', col='grey') +
        theme(legend.position = "bottom") +
	geom_text(x=20, y=0, label=paste0("R2=",round(test$rsq,2)), size=1.5) +
	geom_text(x=20, y=0.1, label=paste0("t1/2=",round(test$halflife,2)), size=1.5)

test <- subset(normData, ID %in% sample_ids$geneID ) %>%
	mutate(rsq = out2$rsq[match(interaction(ID,Condition), interaction(out2$geneID,out2$Condition))]) %>% 
	mutate(halflife=out2$halflife[match(interaction(ID,Condition), interaction(out2$geneID,out2$Condition))])

ggplot(data=test, aes(x = Timepoint, y = decaynorm_frac_cpm, col = Condition)) +
        geom_point(size=0.4, alpha=0.4) +
	geom_smooth(method="lm", formula = y ~ x, se=F, alpha = 0.8, linetype='dashed') +
        facet_grid(Condition~ID) +
        scale_y_continuous(name = "normalized relative abundance") +
        scale_x_continuous(name = "Time post-infiltration (min)", breaks=c(10,20,30,40)) +
        theme_niwot() +
        theme(legend.position = "bottom") +
	geom_hline(yintercept = 0.5, linetype='dashed', col='grey') +
        geom_text(x=30, y=0.5, label=paste0("R2=",round(test$rsq,2)), size=1.5) +
        geom_text(x=30, y=0.8, label=paste0("t1/2=",round(test$halflife,2)), size=1.5)
dev.off()

##########################
## Compute half-life differences between studies
##########################

###
### Between studies NL

b1 <- select(out_NL, geneID, halflife, Condition)

## read in other datasets
b2 <- read_csv("Narsai_2007_mRNA_half_lives.csv", skip=5) %>%
	clean_names() %>%
	select(c(8,13)) %>%
	mutate(geneID = sapply(substr(atnum, 1, 9), function(l) l)) %>%
	group_by(geneID) %>%
	summarise(halflife = median(t1_2_hours) * 60) %>%
	filter( halflife < 24*30*3*60) %>% ## filter half-life < 3 months
	ungroup() %>%
	mutate(geneID = sub(geneID, pattern = 't', replacement = 'T')) %>%
	mutate(geneID = sub(geneID, pattern = 'g', replacement = 'G' )) %>%
	select(geneID, halflife) %>% 
	mutate(Condition = "Narsai")

b3 <- read_csv("Sorenson_2018_mRNA_decay.csv", skip=4) %>% clean_names() %>%
	select(c(1:4)) %>%
	dplyr::rename(geneID = agi) %>%
	mutate(halflife = log(2)/alpha_sov) %>%
	filter(halflife != "Inf") %>%
	filter(halflife < 24*30*3*60) %>% ## filter half-life < 3 months
	select(geneID, halflife) %>% 
	mutate(Condition = "Sorenson")

b4 <- read_csv("TPC2019-LSB-00214R3_Supplemental_Data_Set_5_exponential.csv", skip=4, col_names = T) %>%
	clean_names() %>%
	dplyr::rename(geneID = gene_id) %>%
	mutate(halflife= self_half_life_x*60) %>%
	filter(halflife < 24*30*3*60) %>% ## filter half-life < 3 months
	select(geneID, halflife) %>% 
	mutate(Condition = "Szabo")

b5 <- read_csv("41477_2020_681_MOESM3_ESM_TS4.csv") %>%
	clean_names() %>%
	dplyr::rename(geneID = gene) %>%
	mutate(halflife = log(2)/alpha_wt) %>%
	select(geneID,halflife) %>%
	mutate(Condition = "Chantarachot") 

common_genes <- Reduce(intersect, list(b1$geneID, b2$geneID, b3$geneID, b4$geneID, b5$geneID))

w1 <- rbind(b1,b2,b3,b4,b5) %>%
	subset(geneID %in% common_genes) %>%
	mutate(Condition = gsub("NL","Smith",Condition)) %>%
	mutate(Condition = factor(Condition, levels = c("Smith","Narsai", "Sorenson", "Szabo", "Chantarachot"))) %>%
	mutate(geneID = as.character(geneID)) %>%
	mutate(log_halflife = log2(halflife + 0.01))

pdf("scatterplot_matrix.pdf")

select(w1,-halflife ) %>% spread(Condition, log_halflife) %>% gather(dataset, halflife, -geneID, -Smith) %>%
	ggplot(.) + aes(x=Smith, y=halflife) +
		geom_point() +
		theme_niwot()+
		scale_x_continuous(name="log2 halflife", limits=c(0,15)) +
		scale_y_continuous(name="log2 halflife", limits=c(0,15)) +
		facet_wrap(~dataset) +
		geom_abline(col='salmon', linetype='dashed')

dev.off()

test <- dabest(.data = w1, x = Condition, 
	y = halflife, 
	#y = log_halflife,
	idx = c("Smith", "Narsai", "Sorenson", "Szabo", "Chantarachot"), 
	paired=TRUE, id.col = "geneID")

test1 <- cliffs_delta(test, reps=10000)
test2 <- mean_diff(test, reps=10000)
test3 <- median_diff(test, reps=10000)

pdf("dabest_NL_log2.pdf")
#plot(test1, rawplot.ylabel = "log2 half-life + 0.01 (min)")
#plot(test2, rawplot.ylabel = "log2 half-life + 0.01 (min)")

ggplot(data=test1$data) + aes(x=Condition, y=log_halflife, fill=Condition) +
	geom_boxplot() +
	stat_boxplot(geom="errorbar", width=0.5) +
	theme_niwot() +
	scale_y_continuous(name="log2 half-life + 0.01 (min)") +
	scale_x_discrete(name="Dataset") +
	annotate("text", parse=TRUE,
		x=c(test3$result$control_group[1], test3$result$test_group),
		y=max(test3$data$log_halflife)*1.05,
		label=paste0("theta=", round(c(0,test3$result$difference),2))) +
	annotate("text", parse=TRUE,
		x=c(test1$result$control_group[1], test1$result$test_group), 
		y=max(test1$data$log_halflife)*1.1,
		label=paste("delta",round(c(0,test1$result$difference),2), sep='='))
dev.off()

a1 <- data.frame(test = test3$result$test_group, diff = test3$result$difference)
a <- select(w1, geneID, Condition, halflife) %>% 
	spread(Condition, halflife) %>%
	gather(group, halflife, -geneID, -Smith) %>%
	mutate(diff = a1$diff[match(group, a1$test)]) %>%
	mutate(x = halflife - Smith) %>%
	mutate(test = ifelse(x > diff, yes = 1, no = 0))

a_venn <- select(a, geneID, group, test) %>% spread(group, test)

a2 <- data.frame(geneID = a_venn$geneID, overlaps = rowSums(a_venn[2:5])) %>%
	subset(overlaps > 2)

a3 <- select(a2, geneID) %>% 
	mutate(Smith = a$Smith[match(geneID, a$geneID)]) %>%
	mutate(Narsai = b2$halflife[match(geneID, b2$geneID)]) %>%
	mutate(Sorenson = b3$halflife[match(geneID, b3$geneID)]) %>%
	mutate(Szabo = b4$halflife[match(geneID, b4$geneID)]) %>%
	mutate(Chantarachot = b5$halflife[match(geneID, b5$geneID)]) %>%
	mutate(symbol = anno$name[match(geneID, anno$ID)]) %>%
	mutate(description = anno$description[match(geneID, anno$ID)])

write_csv(a3, "Smith_US_low_halflife.csv")

pdf("venn_Smith_low-halflife.pdf")
vennDiagram(a_venn[2:5], mar=c(2,2,2,2), lwd=1, cex=1)
dev.off()

################
### Compute half-life difference between conditions
###############

#testx <- subset(model_ids, decay_cat == 4) %>% left_join(out_all)

test <- dabest(.data = out_all, #testx, 
	x = Condition,
        y = halflife,
        idx = c("NL", "EL", "Rec"),
        #idx = c("NL","EL"),
	paired=TRUE, id.col = "geneID")

test1 <- cliffs_delta(test, reps=10000)
test2 <- mean_diff(test, reps=10000)
test3 <- median_diff(test, reps=10000)

pdf("dabest_all_log2.pdf")
plot(test2, slopegraph = TRUE)
plot(test3, slopegraph = TRUE)

ggplot(data=test1$data) + aes(x=Condition, y=log_halflife, fill=Condition) +
 	geom_violin() +
	geom_boxplot() +
        stat_boxplot(geom="errorbar", width=0.5) +
        theme_niwot() +
	scale_y_continuous(name="log2 half-life + 0.01 (min)") +
        scale_x_discrete(name="Dataset") +
        annotate("text", parse=TRUE, 
		x=c(test2$result$control_group[1], test2$result$test_group),
		y=max(test2$data$log_halflife)*1.1,
		label=paste0("theta=", round(c(0,test2$result$difference),2))) +
	annotate("text", parse=TRUE,
                x=c(test1$result$control_group[1], test1$result$test_group),
                y=max(test1$data$log_halflife)*1.05,
                label=paste("delta",round(c(0,test1$result$difference),2), sep='='))
dev.off()

a1 <- data.frame(test = test3$result$test_group, diff = test3$result$difference)
a <- select(out_all, geneID, Condition, halflife) %>%
        spread(Condition, halflife) %>%
        gather(group, halflife, -geneID, -NL) %>%
        mutate(diff = a1$diff[match(group, a1$test)]) %>%
        mutate(x = halflife - NL) %>%
        mutate(test = ifelse(x < diff, yes = -1, no = ifelse(x > 1, 1, 0)))

a_venn <- subset(a, group == "EL") %>% select(geneID, group, test) %>% spread(group, test)

##################
## mRNA half-life vs mRNA fold-change
#################

fit <- glmQLFit(y2, design, robust=TRUE)

my.contrasts <- makeContrasts(
	mock_ELvsNL = Mock_10_EL - Mock_10_NL,
	mock_RECvsEL = Mock_20_Rec - Mock_10_EL,
	mock_RECvsNL = Mock_20_Rec - Mock_10_NL,
	levels=design
)

dge2 <- NULL

for(i in colnames(my.contrasts)){
  res <- glmQLFTest(fit, contrast = my.contrasts[,i])
  tt <- topTags(res, adjust.method = "fdr", sort.by = "none", p.value=1, n=dim(res)[1])$table
  tt$contrast <- paste(i)
  tt$ID <- rownames(tt)
  tt <- select(tt, ID, logCPM, contrast, logFC, FDR)
  dge2 <- rbind(dge2,tt)
}

dge2_HL <- subset(dge2, contrast == "mock_ELvsNL")
dge2_Rec <- subset(dge2, contrast == "mock_RECvsEL")
dge2_Rec_US <- subset(dge2, contrast == "mock_RECvsNL")

half_ratio <- out_all %>% 
	select(geneID, halflife, Condition) %>% 
	spread(Condition, halflife) %>%
	mutate(t_ratio_HL = EL/NL) %>%
	mutate(t_ratio_Rec = Rec/EL) %>%
	mutate(t_ratio_Rec_US = Rec/NL) %>%
	mutate(logfc_HL = dge2_HL$logFC[match(geneID,dge2_HL$ID)]) %>%
	mutate(logfc_Rec = dge2_Rec$logFC[match(geneID,dge2_Rec$ID)]) %>%
	mutate(logfc_Rec_US = dge2_Rec_US$logFC[match(geneID,dge2_Rec$ID)]) %>%
	mutate(logcpm_HL = dge2_HL$logCPM[match(geneID,dge2_HL$ID)]) %>%
	mutate(logcpm_Rec = dge2_Rec$logCPM[match(geneID,dge2_Rec$ID)]) %>%
	mutate(fold_change_HL = 2^logfc_HL) %>%
	mutate(fold_change_Rec = 2^logfc_Rec) %>%
	mutate(fold_change_Rec_US = 2^logfc_Rec_US) %>%
	mutate(loci = rrgd$rrgd_loci[match(geneID, rrgd$locus_identifier)]) %>%
	mutate(loci = ifelse(is.na(loci) == T, yes="non", no="RRGD")) %>%
	mutate(loci = factor(loci, levels = c("non","RRGD")))

out_stab <- select(half_ratio, -NL, -EL, -Rec) %>% subset(t_ratio_HL > 1)

write_csv(out_stab, "Smith_EL_high_halflife.csv")

out <- select(half_ratio, geneID, logcpm_HL, t_ratio_HL, fold_changeo_HL, logcpm_Rec, t_ratio_Rec, fold_change_Rec, loci) %>%
	mutate(test = a_venn$EL[match(geneID, a_venn$geneID)]) %>%
	subset(test == -1) %>%
	mutate(symbol = anno$name[match(geneID, anno$ID)]) %>%
	mutate(description = anno$description[match(geneID, anno$ID)])

select(out, -test)%>% write_csv("Smith_EL_low_halflife.csv")

pdf("mRNA-FC_vs_halflife-ratio.pdf")
p1 <- ggplot(data=half_ratio) + aes(x=t_ratio_HL, y=fold_change_HL) + 
	theme_niwot() +
	geom_point(aes(colour=loci, alpha=loci)) +
	geom_smooth(method='lm', formula = y~x) + 
	scale_y_continuous(name = "mRNA relative abundance", trans = "log10") +
        scale_x_continuous(name = "mRNA half-life ratio", trans = "log10") +
	scale_colour_manual(values = c("black", "salmon")) +
	scale_alpha_manual(values = c(0.6,1.4)) +
	geom_vline(xintercept=1, col = 'red', linetype = 'dashed')

p2 <- ggplot(data=half_ratio) + aes(x=logcpm_HL, y=t_ratio_HL) + 
	theme_niwot() +
        geom_point() +
        scale_y_continuous(name = "mRNA half-life ratio") +
        scale_x_continuous(name = "mRNA abundance")

p3 <- ggplot() + 
       	theme_niwot() +
        geom_point(data=subset(half_ratio, loci == "non"), colour='black', alpha=0.6, mapping=aes(x=t_ratio_Rec, y=fold_change_Rec)) +
	geom_point(data=subset(half_ratio, loci == "RRGD"), colour='salmon', alpha=1.6,  mapping=aes(x=t_ratio_Rec, y=fold_change_Rec)) +
	geom_smooth(data = half_ratio, mapping = aes(x=t_ratio_Rec, y=fold_change_Rec), method='lm', formula = y~x) +
        scale_y_continuous(name = "mRNA relative abundance", trans = "log10") +
       	scale_x_continuous(name = "mRNA half-life ratio", trans = "log10") +
        geom_vline(xintercept=1, col = 'red', linetype = 'dashed')

p4 <- ggplot() +
        theme_niwot() +
        geom_point(data=subset(half_ratio, loci == "non"), colour='black', alpha=0.6, mapping=aes(x=t_ratio_Rec_US, y=fold_change_Rec_US)) +
        geom_point(data=subset(half_ratio, loci == "RRGD"), colour='salmon', alpha=1.6,  mapping=aes(x=t_ratio_Rec_US, y=fold_change_Rec_US)) +
        geom_smooth(data = half_ratio, mapping = aes(x=t_ratio_Rec_US, y=fold_change_Rec_US), method='lm', formula = y~x) +
        scale_y_continuous(name = "mRNA relative abundance", trans = "log10") +
        scale_x_continuous(name = "mRNA half-life ratio", trans = "log10") +
        geom_vline(xintercept=1, col = 'red', linetype = 'dashed')

p2

p1

p3

p4

dev.off()

summary(lm(formula = fold_change_HL ~ t_ratio_HL, data = half_ratio))
cor(half_ratio$fold_change_HL, half_ratio$t_ratio_HL)

summary(lm(formula = fold_change_Rec ~ t_ratio_Rec, data = half_ratio))
cor(half_ratio$fold_change_Rec, half_ratio$t_ratio_Rec)

##########################
## PCA polysome-seq
#######################
meta_poly <- dir("polysome_rna/") %>% tibble() %>%
        mutate(file_contents = map(., ~read_csv(file.path("polysome_rna", .)))) %>%
        unnest(cols=c(file_contents)) %>%
        gather(sample, rpkm, -., -ID, -Length, na.rm = T) %>%
        mutate(type = sapply(strsplit(sample, "_"), function(l) l[3])) %>%
        mutate(time = sapply(strsplit(., "_"), function(l) l[3])) %>%
        mutate(time = paste0('t',time)) %>%
        mutate(sample = sapply(strsplit(sample, "_"), function(l) paste0(l[1],l[2])))

total_countFiles <- dir("../Exp566/total_counts/")
lbls <- sapply(strsplit(total_countFiles, "_"), function(l) l[2])
grps <- paste0("Sample",lbls)

v1 <- readDGE(total_countFiles, columns = c(1,7), 
	labels=lbls, group=grps,
	path="../Exp566/total_counts/", skip=1)
v1$samples$time <- meta_poly$time[match(v1$samples$group, meta_poly$sample)]
time <- v1$samples$time
design <- model.matrix(~0 + time)

keep <- rowSums(cpm(v1) > 1) > 3
v1 <- v1[keep, ]
v1$samples$lib.size <- colSums(v1$counts)
v1 <- calcNormFactors(v1, method = "TMM")

poly_countFiles <- dir("../Exp566/poly_counts/")
lbls <- sapply(strsplit(poly_countFiles, "_"), function(l) l[2])
grps <- paste0("Sample",lbls)

v2 <- readDGE(poly_countFiles, columns = c(1,7),
        labels=lbls, group=grps,
        path="../Exp566/poly_counts/", skip=1)
v2$samples$time <- meta_poly$time[match(v2$samples$group, meta_poly$sample)]
time <- v2$samples$time
design <- model.matrix(~0 + time)

keep <- rowSums(cpm(v2) > 1) > 3
v2 <- v2[keep, ]
v2$samples$lib.size <- colSums(v2$counts)
v2 <- calcNormFactors(v2, method = "TMM")

pdf("PCAplots_polysomeRNA.pdf")
mds <- plotMDS(v1, ndim=3, plot=F, gene.selection = "pairwise")

mds_df <- data.frame(Sample = colnames(mds$distance.matrix.squared),
        x1 = as.numeric(mds$x), y1 = as.numeric(mds$y)) %>%
        mutate(Sample = paste0("Sample",Sample)) %>%
        mutate(time = meta_poly$time[match(Sample, meta_poly$sample)])

p1 <- ggplot(data = mds_df) +
        aes(x=x1, y=y1, colour=time) +
        geom_point() +
        ggtitle("Total RNA") +
	scale_x_continuous(name = "Dim1", breaks = c(-1, 0, 1)) +
        scale_y_continuous(name = "Dim2", breaks = c(-1, 0, 1)) +
        theme_niwot() +
	theme(legend.position="none")

mds <- plotMDS(v2, ndim=3, plot=F, gene.selection = "pairwise")

mds_df <- data.frame(Sample = colnames(mds$distance.matrix.squared),
        x1 = as.numeric(mds$x), y1 = as.numeric(mds$y)) %>%
        mutate(Sample = paste0("Sample",Sample)) %>%
        mutate(time = meta_poly$time[match(Sample, meta_poly$sample)])

p2 <- ggplot(data = mds_df) +
        aes(x=x1, y=y1, colour=time) +
        geom_point() +
        ggtitle("Polysome RNA") +
        scale_x_continuous(name = "Dim1", breaks = c(-1, 0, 1)) +
        scale_y_continuous(name = "Dim2", breaks = c(-1, 0, 1)) +
        theme_niwot() + 
	theme(legend.position="none")

p1 + p2 + p1 + p2 + theme(legend.position="bottom")

dev.off()

##################
## mRNA half-life vs translation (polysome-RNA abundance and RPL)
#################
polyR <-meta_poly %>%
	group_by(ID,type,time) %>% summarise(avg_rpkm = mean(rpkm)) %>%
	subset(time == 't0' | time == 't30' | time == 't60' | time == 't90') %>%
	spread(type, avg_rpkm) %>%
	mutate(rpl = poly/total) %>%
	gather(type, rpkm, -ID, -time) %>%
	subset(type == "rpl") %>%
	spread(time, rpkm) %>%
	mutate(fc_30 = t30/t0, fc_60 = t60/t30, fc_90 = t90/t60) %>%
	mutate(logfc_30 = ifelse(fc_30 < 1, -log2(1/fc_30), log2(fc_30))) %>%
	mutate(logfc_60 = ifelse(fc_60 < 1, -log2(1/fc_60), log2(fc_60))) %>%
	mutate(logfc_90 = ifelse(fc_90 < 1, -log2(1/fc_90), log2(fc_90))) %>%
	mutate(t_ratio_HL = half_ratio$t_ratio_HL[match(ID, half_ratio$geneID)]) %>%
	mutate(log_half_HL = ifelse(t_ratio_HL < 1, -log2(1/t_ratio_HL), log2(t_ratio_HL))) %>%
	mutate(t_ratio_REC = half_ratio$t_ratio_REC[match(ID, half_ratio$geneID)]) %>%
	mutate(log_half_REC = ifelse(t_ratio_REC<1, -log2(1/t_ratio_REC), log2(t_ratio_REC))) %>%
	na.omit()  %>%
        mutate(loci = rrgd$rrgd_loci[match(ID, rrgd$locus_identifier)]) %>%
        mutate(loci = ifelse(is.na(loci) == T, yes="non", no="RRGD")) %>%
        mutate(loci = factor(loci, levels = c("non","RRGD"))) %>%
	mutate(cat_30 = ifelse(logfc_30 >= log2(1.5), 'g1', ifelse(logfc_30 <= -log2(1.5), 'g2', 'g3'))) %>% 
	mutate(cat_30 = ifelse(cat_30 != 'g3', cat_30, ifelse(log_half_HL>=log2(1.5),'g4', ifelse(log_half_HL <= -log2(1.5),'g5','g6')))) %>%
	mutate(cat_60 = ifelse(logfc_60 >= log2(1.5), 'g1', ifelse(logfc_60 <= -log2(1.5), 'g2', 'g3'))) %>%
        mutate(cat_60 = ifelse(cat_60 != 'g3', cat_60, ifelse(log_half_HL>=log2(1.5),'g4',ifelse(log_half_HL <= -log2(1.5),'g5','g6')))) %>%
	mutate(cat_90 = ifelse(logfc_90 >= log2(1.5), 'g1', ifelse(logfc_90 <= -log2(1.5), 'g2', 'g3'))) %>%
        mutate(cat_90 = ifelse(cat_90 != 'g3', cat_90,ifelse(log_half_REC>=log2(1.5),'g4',ifelse(log_half_REC <= -log2(1.5),'g5','g6'))))

pdf("rpl_vs_halflife.pdf", paper='a4')
p1 <- ggplot(data=polyR) + aes(x=log_half_HL, y=logfc_30, col = cat_30) +
	theme_bw() +
        geom_point(size=0.2) +
        scale_y_continuous(name = "log2FC RPL", breaks=c(-5,0,5)) +
	geom_hline(yintercept = c(log2(1.5),-log2(1.5)), linetype = 'dashed') +
        scale_x_continuous(name = "log2FC half-life ratio \n (HL : US)", breaks=c(-5,0,5)) +
	geom_vline(xintercept = c(log2(1.5),-log2(1.5)), linetype = 'dashed') +
        scale_colour_manual(values = c("red2","steelblue2","salmon","skyblue","grey")) +
	theme(legend.position="none") +
	coord_cartesian(ylim=c(-7.5,7.5), xlim=c(-7.5,7.5))

p2 <- ggplot(data=polyR) + aes(x=log_half_HL, y=logfc_60, col = cat_60) +
        theme_bw() +
        geom_point(size=0.2) +
        scale_y_continuous(name = "log2FC RPL", breaks=c(-5,0,5)) +
        scale_x_continuous(name = "log2FC half-life ratio \n (HL : US)", breaks=c(-5,0,5)) +
	geom_vline(xintercept = c(log2(1.5),-log2(1.5)), linetype = 'dashed') +
	geom_hline(yintercept = c(log2(1.5),-log2(1.5)), linetype = 'dashed') +
        scale_colour_manual(values = c("red2","steelblue2","salmon","skyblue","grey")) +
	theme(legend.position="none") +
	coord_cartesian(ylim=c(-7.5,7.5), xlim=c(-7.5,7.5))

p3 <- ggplot(data=polyR) + aes(x=log_half_REC, y=logfc_90, col=cat_90) +
        theme_bw() +
        geom_point(size=0.2) +
        scale_y_continuous(name = "log2FC RPL", breaks=c(-5,0,5)) +
        scale_x_continuous(name = "log2FC half-life ratio \n (REC : HL)", breaks=c(-5,0,5)) +
	geom_vline(xintercept = c(log2(1.5),-log2(1.5)), linetype = 'dashed') +
        geom_hline(yintercept = c(log2(1.5),-log2(1.5)), linetype = 'dashed') +
        scale_colour_manual(values = c("red2","steelblue2","salmon","skyblue","grey")) +
        theme(legend.position="none")+
        coord_cartesian(ylim=c(-7.5,7.5), xlim=c(-7.5,7.5))

p4 <- ggplot() + geom_point()

(p1 + p2 + p3) / (p4 + p4 + p4) / (p4 + p4 + p4)

dev.off()

## write out genes with increased RPL at 30-60 HL
test_rpl <- subset(polyR, cat_60 == "g1") %>%
	select(ID, logfc_60, t_ratio_HL, log_half_HL, cat_60) %>%
	arrange(desc(log2fc_60))
#subset(log_half_HL <= log(1.5) & log_half_HL >= -log(1.5)) ## no change in stability

write_csv(test_rpl, "HL_60_RPLup.csv")

#### collate Poly and total RPKM
poly_rpkm <- meta_poly %>%
	group_by(ID,type,time) %>% summarise(avg_rpkm = mean(rpkm)) %>%
	mutate(sample = paste(type, time, sep="_")) %>%
	ungroup() %>% dplyr::select(-type, -time) %>%
        spread(sample, avg_rpkm)

## abundance filter: RPKM > 1 in >3 samples
polyR_rpkm[is.na(polyR_rpkm)] <- 0
keep <- rowSums(polyR_rpkm>1)>3
polyR_rpkm <- polyR_rpkm[keep,]

write_csv(polyR_rpkm, "Polysome_rpkm.csv")



